{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ef9f43",
   "metadata": {},
   "outputs": [],
   "source": [
    "## use RNA-STAR conda environment\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed68539",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "704f4313",
   "metadata": {},
   "outputs": [],
   "source": [
    "left = pd.read_csv(\"UNUAR_motif_sites_mRNA.tsv\", sep = \"\\t\")\n",
    "left[\"Motif\"] = left[\"Motif\"].str.replace(\"U\", \"T\") ## each BAM file has sequences w/ Thymine (T) instead of Uracil (U)\n",
    "right = pd.read_excel(\"SupplementaryTable1.xlsx\")\n",
    "df = pd.merge(left, right, how = \"left\", on = \"Motif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e607fdf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TranscriptID', 'Motif', 'Region', 'Chrom', 'Strand',\n",
       "       'TranscriptPosStart', 'TranscriptPosEnd', 'TranscriptModBase',\n",
       "       'GenomicModBase', 'TranscriptLength', 'DistFromAUG', 'DistFromSTOP',\n",
       "       'DistFromExonStart', 'DistFromExonEnd', 'fit_A', 'fit_B', 'fit_R'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2c0ab995",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>DistFromAUG</th>\n",
       "      <th>DistFromSTOP</th>\n",
       "      <th>DistFromExonStart</th>\n",
       "      <th>DistFromExonEnd</th>\n",
       "      <th>fit_A</th>\n",
       "      <th>fit_B</th>\n",
       "      <th>fit_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3425688</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TATAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2262</td>\n",
       "      <td>2267</td>\n",
       "      <td>2264</td>\n",
       "      <td>248918727</td>\n",
       "      <td>2684</td>\n",
       "      <td>2133</td>\n",
       "      <td>364</td>\n",
       "      <td>2125</td>\n",
       "      <td>419</td>\n",
       "      <td>0.829824</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>0.684980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425689</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TCTAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2267</td>\n",
       "      <td>2272</td>\n",
       "      <td>2269</td>\n",
       "      <td>248918732</td>\n",
       "      <td>2684</td>\n",
       "      <td>2138</td>\n",
       "      <td>369</td>\n",
       "      <td>2130</td>\n",
       "      <td>414</td>\n",
       "      <td>0.753295</td>\n",
       "      <td>3.700348e-02</td>\n",
       "      <td>0.763335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425690</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TATAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2286</td>\n",
       "      <td>2291</td>\n",
       "      <td>2288</td>\n",
       "      <td>248918751</td>\n",
       "      <td>2684</td>\n",
       "      <td>2157</td>\n",
       "      <td>388</td>\n",
       "      <td>2149</td>\n",
       "      <td>395</td>\n",
       "      <td>0.829824</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>0.684980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425691</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TGTAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2297</td>\n",
       "      <td>2302</td>\n",
       "      <td>2299</td>\n",
       "      <td>248918762</td>\n",
       "      <td>2684</td>\n",
       "      <td>2168</td>\n",
       "      <td>399</td>\n",
       "      <td>2160</td>\n",
       "      <td>384</td>\n",
       "      <td>0.763066</td>\n",
       "      <td>3.536472e-02</td>\n",
       "      <td>0.670875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425692</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TTTAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2380</td>\n",
       "      <td>2385</td>\n",
       "      <td>2382</td>\n",
       "      <td>248918845</td>\n",
       "      <td>2684</td>\n",
       "      <td>2251</td>\n",
       "      <td>482</td>\n",
       "      <td>2243</td>\n",
       "      <td>301</td>\n",
       "      <td>0.841842</td>\n",
       "      <td>8.050644e-02</td>\n",
       "      <td>0.610735</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               TranscriptID  Motif Region         Chrom Strand  \\\n",
       "3425688  rna-XM_047417842.1  TATAA   3UTR  NC_000001.11      +   \n",
       "3425689  rna-XM_047417842.1  TCTAA   3UTR  NC_000001.11      +   \n",
       "3425690  rna-XM_047417842.1  TATAA   3UTR  NC_000001.11      +   \n",
       "3425691  rna-XM_047417842.1  TGTAA   3UTR  NC_000001.11      +   \n",
       "3425692  rna-XM_047417842.1  TTTAG   3UTR  NC_000001.11      +   \n",
       "\n",
       "         TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  \\\n",
       "3425688                2262              2267               2264   \n",
       "3425689                2267              2272               2269   \n",
       "3425690                2286              2291               2288   \n",
       "3425691                2297              2302               2299   \n",
       "3425692                2380              2385               2382   \n",
       "\n",
       "         GenomicModBase  TranscriptLength  DistFromAUG  DistFromSTOP  \\\n",
       "3425688       248918727              2684         2133           364   \n",
       "3425689       248918732              2684         2138           369   \n",
       "3425690       248918751              2684         2157           388   \n",
       "3425691       248918762              2684         2168           399   \n",
       "3425692       248918845              2684         2251           482   \n",
       "\n",
       "         DistFromExonStart  DistFromExonEnd     fit_A         fit_B     fit_R  \n",
       "3425688               2125              419  0.829824  1.000000e-08  0.684980  \n",
       "3425689               2130              414  0.753295  3.700348e-02  0.763335  \n",
       "3425690               2149              395  0.829824  1.000000e-08  0.684980  \n",
       "3425691               2160              384  0.763066  3.536472e-02  0.670875  \n",
       "3425692               2243              301  0.841842  8.050644e-02  0.610735  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9772399b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null values: 0\n"
     ]
    }
   ],
   "source": [
    "null_val = df.isnull().sum().sum()\n",
    "print(\"Null values: \" + str(null_val))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a740903",
   "metadata": {},
   "source": [
    "### Opening BAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fdfb7666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>DistFromAUG</th>\n",
       "      <th>DistFromSTOP</th>\n",
       "      <th>DistFromExonStart</th>\n",
       "      <th>DistFromExonEnd</th>\n",
       "      <th>fit_A</th>\n",
       "      <th>fit_B</th>\n",
       "      <th>fit_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_001134855.2-2</td>\n",
       "      <td>TCTAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NW_018654708.1</td>\n",
       "      <td>-</td>\n",
       "      <td>1358</td>\n",
       "      <td>1363</td>\n",
       "      <td>1360</td>\n",
       "      <td>373</td>\n",
       "      <td>1733</td>\n",
       "      <td>880</td>\n",
       "      <td>151</td>\n",
       "      <td>101</td>\n",
       "      <td>372</td>\n",
       "      <td>0.716983</td>\n",
       "      <td>5.129877e-02</td>\n",
       "      <td>0.678453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_001134855.2-2</td>\n",
       "      <td>TCTAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_018654708.1</td>\n",
       "      <td>-</td>\n",
       "      <td>1530</td>\n",
       "      <td>1535</td>\n",
       "      <td>1532</td>\n",
       "      <td>201</td>\n",
       "      <td>1733</td>\n",
       "      <td>1052</td>\n",
       "      <td>21</td>\n",
       "      <td>273</td>\n",
       "      <td>200</td>\n",
       "      <td>0.753295</td>\n",
       "      <td>3.700348e-02</td>\n",
       "      <td>0.763335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_001171747.2-2</td>\n",
       "      <td>TCTAA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NW_019805492.1</td>\n",
       "      <td>+</td>\n",
       "      <td>98</td>\n",
       "      <td>103</td>\n",
       "      <td>100</td>\n",
       "      <td>101</td>\n",
       "      <td>1500</td>\n",
       "      <td>100</td>\n",
       "      <td>187</td>\n",
       "      <td>100</td>\n",
       "      <td>1399</td>\n",
       "      <td>0.753295</td>\n",
       "      <td>3.700348e-02</td>\n",
       "      <td>0.763335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001171747.2-2</td>\n",
       "      <td>TTTAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_019805492.1</td>\n",
       "      <td>+</td>\n",
       "      <td>357</td>\n",
       "      <td>362</td>\n",
       "      <td>359</td>\n",
       "      <td>360</td>\n",
       "      <td>1500</td>\n",
       "      <td>359</td>\n",
       "      <td>72</td>\n",
       "      <td>359</td>\n",
       "      <td>1140</td>\n",
       "      <td>0.841842</td>\n",
       "      <td>8.050644e-02</td>\n",
       "      <td>0.610735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_001171747.2-2</td>\n",
       "      <td>TTTAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_019805492.1</td>\n",
       "      <td>+</td>\n",
       "      <td>383</td>\n",
       "      <td>388</td>\n",
       "      <td>385</td>\n",
       "      <td>386</td>\n",
       "      <td>1500</td>\n",
       "      <td>385</td>\n",
       "      <td>98</td>\n",
       "      <td>385</td>\n",
       "      <td>1114</td>\n",
       "      <td>0.841842</td>\n",
       "      <td>8.050644e-02</td>\n",
       "      <td>0.610735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425688</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TATAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2262</td>\n",
       "      <td>2267</td>\n",
       "      <td>2264</td>\n",
       "      <td>248918727</td>\n",
       "      <td>2684</td>\n",
       "      <td>2133</td>\n",
       "      <td>364</td>\n",
       "      <td>2125</td>\n",
       "      <td>419</td>\n",
       "      <td>0.829824</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>0.684980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425689</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TCTAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2267</td>\n",
       "      <td>2272</td>\n",
       "      <td>2269</td>\n",
       "      <td>248918732</td>\n",
       "      <td>2684</td>\n",
       "      <td>2138</td>\n",
       "      <td>369</td>\n",
       "      <td>2130</td>\n",
       "      <td>414</td>\n",
       "      <td>0.753295</td>\n",
       "      <td>3.700348e-02</td>\n",
       "      <td>0.763335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425690</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TATAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2286</td>\n",
       "      <td>2291</td>\n",
       "      <td>2288</td>\n",
       "      <td>248918751</td>\n",
       "      <td>2684</td>\n",
       "      <td>2157</td>\n",
       "      <td>388</td>\n",
       "      <td>2149</td>\n",
       "      <td>395</td>\n",
       "      <td>0.829824</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>0.684980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425691</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TGTAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2297</td>\n",
       "      <td>2302</td>\n",
       "      <td>2299</td>\n",
       "      <td>248918762</td>\n",
       "      <td>2684</td>\n",
       "      <td>2168</td>\n",
       "      <td>399</td>\n",
       "      <td>2160</td>\n",
       "      <td>384</td>\n",
       "      <td>0.763066</td>\n",
       "      <td>3.536472e-02</td>\n",
       "      <td>0.670875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425692</th>\n",
       "      <td>rna-XM_047417842.1</td>\n",
       "      <td>TTTAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>+</td>\n",
       "      <td>2380</td>\n",
       "      <td>2385</td>\n",
       "      <td>2382</td>\n",
       "      <td>248918845</td>\n",
       "      <td>2684</td>\n",
       "      <td>2251</td>\n",
       "      <td>482</td>\n",
       "      <td>2243</td>\n",
       "      <td>301</td>\n",
       "      <td>0.841842</td>\n",
       "      <td>8.050644e-02</td>\n",
       "      <td>0.610735</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3425693 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TranscriptID  Motif Region           Chrom Strand  \\\n",
       "0        rna-NM_001134855.2-2  TCTAG    CDS  NW_018654708.1      -   \n",
       "1        rna-NM_001134855.2-2  TCTAA   3UTR  NW_018654708.1      -   \n",
       "2        rna-NM_001171747.2-2  TCTAA    CDS  NW_019805492.1      +   \n",
       "3        rna-NM_001171747.2-2  TTTAG   3UTR  NW_019805492.1      +   \n",
       "4        rna-NM_001171747.2-2  TTTAG   3UTR  NW_019805492.1      +   \n",
       "...                       ...    ...    ...             ...    ...   \n",
       "3425688    rna-XM_047417842.1  TATAA   3UTR    NC_000001.11      +   \n",
       "3425689    rna-XM_047417842.1  TCTAA   3UTR    NC_000001.11      +   \n",
       "3425690    rna-XM_047417842.1  TATAA   3UTR    NC_000001.11      +   \n",
       "3425691    rna-XM_047417842.1  TGTAA   3UTR    NC_000001.11      +   \n",
       "3425692    rna-XM_047417842.1  TTTAG   3UTR    NC_000001.11      +   \n",
       "\n",
       "         TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  \\\n",
       "0                      1358              1363               1360   \n",
       "1                      1530              1535               1532   \n",
       "2                        98               103                100   \n",
       "3                       357               362                359   \n",
       "4                       383               388                385   \n",
       "...                     ...               ...                ...   \n",
       "3425688                2262              2267               2264   \n",
       "3425689                2267              2272               2269   \n",
       "3425690                2286              2291               2288   \n",
       "3425691                2297              2302               2299   \n",
       "3425692                2380              2385               2382   \n",
       "\n",
       "         GenomicModBase  TranscriptLength  DistFromAUG  DistFromSTOP  \\\n",
       "0                   373              1733          880           151   \n",
       "1                   201              1733         1052            21   \n",
       "2                   101              1500          100           187   \n",
       "3                   360              1500          359            72   \n",
       "4                   386              1500          385            98   \n",
       "...                 ...               ...          ...           ...   \n",
       "3425688       248918727              2684         2133           364   \n",
       "3425689       248918732              2684         2138           369   \n",
       "3425690       248918751              2684         2157           388   \n",
       "3425691       248918762              2684         2168           399   \n",
       "3425692       248918845              2684         2251           482   \n",
       "\n",
       "         DistFromExonStart  DistFromExonEnd     fit_A         fit_B     fit_R  \n",
       "0                      101              372  0.716983  5.129877e-02  0.678453  \n",
       "1                      273              200  0.753295  3.700348e-02  0.763335  \n",
       "2                      100             1399  0.753295  3.700348e-02  0.763335  \n",
       "3                      359             1140  0.841842  8.050644e-02  0.610735  \n",
       "4                      385             1114  0.841842  8.050644e-02  0.610735  \n",
       "...                    ...              ...       ...           ...       ...  \n",
       "3425688               2125              419  0.829824  1.000000e-08  0.684980  \n",
       "3425689               2130              414  0.753295  3.700348e-02  0.763335  \n",
       "3425690               2149              395  0.829824  1.000000e-08  0.684980  \n",
       "3425691               2160              384  0.763066  3.536472e-02  0.670875  \n",
       "3425692               2243              301  0.841842  8.050644e-02  0.610735  \n",
       "\n",
       "[3425693 rows x 17 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2138ff5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def deletion_rate(input_bam_name):\n",
    "    bamfile = pysam.AlignmentFile(input_bam_name, \"rb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa95dd7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_bam(folder_name):\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/\"realignments\"/folder_name\n",
    "    \n",
    "    try: \n",
    "        for subfolder in input_dir.iterdir():\n",
    "            if subfolder.is_dir():\n",
    "                processed_folder = input_dir/f\"{subfolder.name}_realigned\"\n",
    "                \n",
    "                for bam in subfolder.glob(\"*.bam\"):\n",
    "                    input_bam_name = Path(bam) ## turn string from list back into filepath\n",
    "\n",
    "                    # run_realign(input_bam_name, output_bam_name, fasta_dir, discard)\n",
    "                    # bam_index(output_bam_name)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to calculate deletion rates in {folder_name}: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02112d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Calculates observed and real deletion rates for every read in a BAM file.\")\n",
    "    parser.add_argument(\"--folder_name\", help = \"Name of processed_fastqs folder\", required = True)\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting realignment...\")\n",
    "    open_bam(args.folder_name)\n",
    "    print(\"Realignment finished.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
