{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ef9f43",
   "metadata": {},
   "outputs": [],
   "source": [
    "## use RNA-STAR conda environment\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed68539",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "704f4313",
   "metadata": {},
   "outputs": [],
   "source": [
    "left = pd.read_csv(\"UNUAR_motif_sites_mRNA.tsv\", sep = \"\\t\")\n",
    "left[\"Motif\"] = left[\"Motif\"].str.replace(\"U\", \"T\") ## each BAM file has sequences w/ Thymine (T) instead of Uracil (U)\n",
    "right = pd.read_excel(\"SupplementaryTable1.xlsx\")\n",
    "\n",
    "df = pd.merge(left, right, how = \"left\", on = \"Motif\")\n",
    "genome_coord = df[[\"Chrom\", \"GenomicModBase\"]].to_numpy() ## faster processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a740903",
   "metadata": {},
   "source": [
    "### Opening BAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdfb7666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>DistFromAUG</th>\n",
       "      <th>DistFromSTOP</th>\n",
       "      <th>DistFromExonStart</th>\n",
       "      <th>DistFromExonEnd</th>\n",
       "      <th>fit_A</th>\n",
       "      <th>fit_B</th>\n",
       "      <th>fit_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_001134855.2-2</td>\n",
       "      <td>TCTAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NW_018654708.1</td>\n",
       "      <td>-</td>\n",
       "      <td>1358</td>\n",
       "      <td>1363</td>\n",
       "      <td>1360</td>\n",
       "      <td>373</td>\n",
       "      <td>1733</td>\n",
       "      <td>880</td>\n",
       "      <td>151</td>\n",
       "      <td>101</td>\n",
       "      <td>372</td>\n",
       "      <td>0.716983</td>\n",
       "      <td>0.051299</td>\n",
       "      <td>0.678453</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           TranscriptID  Motif Region           Chrom Strand  \\\n",
       "0  rna-NM_001134855.2-2  TCTAG    CDS  NW_018654708.1      -   \n",
       "\n",
       "   TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  GenomicModBase  \\\n",
       "0                1358              1363               1360             373   \n",
       "\n",
       "   TranscriptLength  DistFromAUG  DistFromSTOP  DistFromExonStart  \\\n",
       "0              1733          880           151                101   \n",
       "\n",
       "   DistFromExonEnd     fit_A     fit_B     fit_R  \n",
       "0              372  0.716983  0.051299  0.678453  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33b5bab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pysam_pileup(bamfile, chrom, mod_base, base_ct):\n",
    "    \"\"\"\n",
    "    Counts number of bases/deletions in PileupColumn \n",
    "    for each GenomicModBase\n",
    "    \"\"\"\n",
    "    try:\n",
    "        for pileupcolumn in bamfile.pileup(str(chrom), int(mod_base-1), int(mod_base), truncate = True): ## pysam is 0-based, while GenomicModBase was 1-based\n",
    "            pileupcolumn.set_min_base_quality(0) ## prevent pysam from filtering based on base quality\n",
    "            if pileupcolumn.pos != int(mod_base-1): ## if not at exact pos, skip to next iteration\n",
    "                continue\n",
    "            for pileupread in pileupcolumn.pileups:\n",
    "                if pileupread.is_del and not pileupread.is_refskip:\n",
    "                    base_ct[\"Deletions\"] += 1\n",
    "                elif not pileupread.is_refskip:\n",
    "                    base = pileupread.alignment.query_sequence[pileupread.query_position] ## taken from example in manual; returns base letter\n",
    "                    base_ct[str(base)] += 1\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to count bases/deletions in PileupColumn for Chromosome {chrom} at GenomicModBase {mod_base}: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise\n",
    "\n",
    "def count_base(input_bam_name, results):\n",
    "    bamfile = pysam.AlignmentFile(input_bam_name, \"rb\")\n",
    "    \"\"\"\n",
    "    Counts number of bases/deletions for each UNUAR site\n",
    "        1) chrom: value in \"Chrom\" column (e.g., NW_018654708.1)\n",
    "        2) mod_base: value in \"GenomicModBase\" column (e.g., 373)\n",
    "    \"\"\"\n",
    "    try:\n",
    "        all_chunks = np.array_split(genome_coord, 100)\n",
    "\n",
    "        for chunk in all_chunks:\n",
    "            for row in chunk:\n",
    "                chrom = row[0]\n",
    "                mod_base = row[1]\n",
    "                base_ct = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0, \"Deletions\": 0}\n",
    "\n",
    "                pysam_pileup(bamfile, chrom, mod_base, base_ct)\n",
    "\n",
    "                results.append({\"Chrom\": chrom,\n",
    "                                \"GenomicModBase\": mod_base,\n",
    "                                **base_ct}) ## unpack base_ct dict in this dict\n",
    "                \n",
    "        bamfile.close()\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to count bases/deletions in UNUAR sites: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa95dd7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## main code\n",
    "def open_bam(folder_name):\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/\"realignments\"/folder_name\n",
    "    \n",
    "    try: \n",
    "        for subfolder in input_dir.iterdir():\n",
    "            if subfolder.is_dir():\n",
    "                processed_folder = input_dir/f\"{subfolder.name}_realigned\"\n",
    "                \n",
    "                for bam in subfolder.glob(\"*.bam\"):\n",
    "                    results = []\n",
    "                    input_bam_name = Path(bam) ## turn string from list back into filepath\n",
    "                    output_tsv_name = processed_folder/f\"{input_bam_name.stem}.tsv\"\n",
    "                    \n",
    "                    ## count A, C, G, T and deletions @ each UNUAR site\n",
    "                    count_base(input_bam_name, results)\n",
    "\n",
    "                    ## calculate observed deletion rates\n",
    "                    counts = pd.DataFrame(results)\n",
    "                    sums = counts[[\"A\", \"C\", \"G\", \"T\", \"Deletions\"]].sum(axis = 1) ## sum across rows\n",
    "                    counts_filtered = counts[sums != 0] ## only keep rows where sum does not equal 0 (prevents division by 0)\n",
    "                    counts_filtered[\"DeletionRate\"] = counts_filtered[\"Deletions\"]/counts_filtered[[\"A\", \"C\", \"G\", \"T\", \"Deletions\"]].sum(axis = 1)\n",
    "                    \n",
    "                    ## calculate real deletion rates\n",
    "                    df_final = pd.merge(df, counts_filtered, how = \"left\", on = [\"Chrom\", \"GenomicModBase\"])\n",
    "                    num = df_final[\"DeletionRate\"] - df_final[\"fit_B\"]\n",
    "                    denom = ((df_final[\"fit_R\"] - df_final[\"fit_B\"]) + \n",
    "                             df_final[\"fit_A\"]*(df_final[\"DeletionRate\"] - df_final[\"fit_R\"] - df_final[\"fit_B\"]))\n",
    "                    df_final[\"RealRate\"] = num/denom\n",
    "\n",
    "                    ## add all calculations to og dataframe & save as .tsv output\n",
    "                    df_final.to_csv(output_tsv_name, sep = \"\\t\", index = False)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to calculate observed & real deletion rates in {folder_name} and save as .tsv: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02112d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Calculates observed and real deletion rates for every UNUAR site in a BAM file.\")\n",
    "    parser.add_argument(\"--folder_name\", help = \"Name of processed_fastqs folder\", required = True)\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Calculating deletion rates...\")\n",
    "    open_bam(args.folder_name)\n",
    "    print(\"Process finished.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
