{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ef9f43",
   "metadata": {},
   "outputs": [],
   "source": [
    "## use RNA-STAR conda environment\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed68539",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "704f4313",
   "metadata": {},
   "outputs": [],
   "source": [
    "left = pd.read_csv(\"UNUAR_motif_sites_mRNA.tsv\", sep = \"\\t\")\n",
    "left[\"Motif\"] = left[\"Motif\"].str.replace(\"U\", \"T\") ## each BAM file has sequences w/ Thymine (T) instead of Uracil (U)\n",
    "right = pd.read_excel(\"SupplementaryTable1.xlsx\")\n",
    "\n",
    "df = pd.merge(left, right, how = \"left\", on = \"Motif\")\n",
    "genome_coord = df[[\"Chrom\", \"GenomicModBase\"]].to_numpy() ## faster processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a740903",
   "metadata": {},
   "source": [
    "### Opening BAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdfb7666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>DistFromAUG</th>\n",
       "      <th>DistFromSTOP</th>\n",
       "      <th>DistFromExonStart</th>\n",
       "      <th>DistFromExonEnd</th>\n",
       "      <th>fit_A</th>\n",
       "      <th>fit_B</th>\n",
       "      <th>fit_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_001134855.2-2</td>\n",
       "      <td>TCTAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NW_018654708.1</td>\n",
       "      <td>-</td>\n",
       "      <td>1358</td>\n",
       "      <td>1363</td>\n",
       "      <td>1360</td>\n",
       "      <td>373</td>\n",
       "      <td>1733</td>\n",
       "      <td>880</td>\n",
       "      <td>151</td>\n",
       "      <td>101</td>\n",
       "      <td>372</td>\n",
       "      <td>0.716983</td>\n",
       "      <td>0.051299</td>\n",
       "      <td>0.678453</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           TranscriptID  Motif Region           Chrom Strand  \\\n",
       "0  rna-NM_001134855.2-2  TCTAG    CDS  NW_018654708.1      -   \n",
       "\n",
       "   TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  GenomicModBase  \\\n",
       "0                1358              1363               1360             373   \n",
       "\n",
       "   TranscriptLength  DistFromAUG  DistFromSTOP  DistFromExonStart  \\\n",
       "0              1733          880           151                101   \n",
       "\n",
       "   DistFromExonEnd     fit_A     fit_B     fit_R  \n",
       "0              372  0.716983  0.051299  0.678453  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e788cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If UNUAR site is at a position where there are 0 reads (btween r1 end, r2 start), then impute n/a into base counts. After all data is appended to TSV, drop all rows with n/a\n",
    "# Append counts to a dictionary, append to dataframe in chunks, then append that dataframe to tsv in chunk\n",
    "# Write (create) columns in tsv before appending dataframe\n",
    "# Ensure that total count == A + G + C + T + Deletions\n",
    "# Add error handling at the end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33b5bab",
   "metadata": {},
   "outputs": [],
   "source": [
    "## my code\n",
    "def pysam_pileup(bamfile, chrom, mod_base, base_counts):\n",
    "    for pileupcolumn in bamfile.pileup(str(chrom), int(mod_base-1), int(mod_base), truncate = True): ## pysam is 0-based, while GenomicModBase was 1-based\n",
    "        pileupcolumn.set_min_base_quality(0) ## prevent pysam from filtering based on base quality\n",
    "        if pileupcolumn.pos != int(mod_base-1): ## if not at exact pos, skip to next iteration\n",
    "            continue\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            if pileupread.is_del and not pileupread.is_refskip:\n",
    "                base_counts[\"del\"] += 1\n",
    "            elif not pileupread.is_refskip:\n",
    "                base = pileupread.alignment.query_sequence[pileupread.query_position] ## taken from example in manual; returns base letter\n",
    "                base_counts[str(base)] += 1\n",
    "\n",
    "def count_base(input_bam_name, fafile):\n",
    "    bamfile = pysam.AlignmentFile(input_bam_name, \"rb\")\n",
    "    results = []\n",
    "    \"\"\"\n",
    "    chrom: value in \"Chrom\" column (e.g., NW_018654708.1)\n",
    "    mod_base: value in \"GenomicModBase\" column (e.g., 373)\n",
    "    \"\"\"\n",
    "    for row in genome_coord:\n",
    "        chrom = row[0]\n",
    "        mod_base = row[1]\n",
    "        base_counts = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0, \"del\": 0}\n",
    "\n",
    "        pysam_pileup(chrom, mod_base, base_counts)\n",
    "        bamfile.close()\n",
    "\n",
    "        results.append({\"Chrom\": chrom,\n",
    "                        \"GenomicModBase\": mod_base,\n",
    "                        **base_counts}) ## unpack base_counts dict in this dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa95dd7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_bam(folder_name, fasta_dir):\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/\"realignments\"/folder_name\n",
    "    fafile = pysam.FastaFile(fasta_dir)\n",
    "    \n",
    "    try: \n",
    "        for subfolder in input_dir.iterdir():\n",
    "            if subfolder.is_dir():\n",
    "                processed_folder = input_dir/f\"{subfolder.name}_realigned\"\n",
    "                \n",
    "                for bam in subfolder.glob(\"*.bam\"):\n",
    "                    input_bam_name = Path(bam) ## turn string from list back into filepath\n",
    "\n",
    "                    count_base(input_bam_name, fasta_dir)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to calculate deletion rates in {folder_name}: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02112d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Calculates observed and real deletion rates for every read in a BAM file.\")\n",
    "    parser.add_argument(\"--folder_name\", help = \"Name of processed_fastqs folder\", required = True)\n",
    "    parser.add_argument(\"--fasta_dir\", help = \"Path of reference fasta file\", required = True)\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting realignment...\")\n",
    "    open_bam(args.folder_name, args.fasta_dir)\n",
    "    print(\"Realignment finished.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d52fa5b7",
   "metadata": {},
   "source": [
    "### Test code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23aac039",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chunk_size = 100000\n",
    "# for i in range(0, len(results), chunk_size):\n",
    "#     chunk = results[i:i+chunk_size]\n",
    "#     df_chunk = pd.DataFrame(chunk)\n",
    "#     df_chunk.to_csv(\"output.tsv\", sep=\"\\t\", index=False,\n",
    "#                     mode=\"a\", header=(i == 0))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
