{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "626ab5ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbf26e5a",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2e45fa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_c</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>DeletionRate</th>\n",
       "      <th>RealRate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_021254.4</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1102</td>\n",
       "      <td>1107</td>\n",
       "      <td>1104</td>\n",
       "      <td>32601765</td>\n",
       "      <td>3516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_001350335.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>2193</td>\n",
       "      <td>2198</td>\n",
       "      <td>2195</td>\n",
       "      <td>32601765</td>\n",
       "      <td>2294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TranscriptID  Motif Region        Chrom Strand  TranscriptPosStart  \\\n",
       "0     rna-NM_021254.4  UGUAA   3UTR  NC_000021.9      -                1102   \n",
       "1  rna-NM_001350335.2  UGUAA   3UTR  NC_000021.9      -                2193   \n",
       "2  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "3  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "4  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "\n",
       "   TranscriptPosEnd  TranscriptModBase  GenomicModBase  TranscriptLength  ...  \\\n",
       "0              1107               1104        32601765              3516  ...   \n",
       "1              2198               2195        32601765              2294  ...   \n",
       "2              1405               1402        32601765              1501  ...   \n",
       "3              1405               1402        32601765              1501  ...   \n",
       "4              1405               1402        32601765              1501  ...   \n",
       "\n",
       "      fit_c     fit_s  fit_b    A    C    G    T  Deletions  DeletionRate  \\\n",
       "0  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "1  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "2  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "3  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "4  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "\n",
       "   RealRate  \n",
       "0  1.058417  \n",
       "1  1.058417  \n",
       "2  1.058417  \n",
       "3  1.058417  \n",
       "4  1.058417  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load in file\n",
    "df = pd.read_csv(\"C:/Users/Sonia Ling/Desktop/calculate_dr/KEH-Rep2-7KO-HEK293T-Cyto-BS.sorted.tsv\", sep = \"\\t\")\n",
    "\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8c6689f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TranscriptID',\n",
       " 'Motif',\n",
       " 'Region',\n",
       " 'Chrom',\n",
       " 'Strand',\n",
       " 'TranscriptPosStart',\n",
       " 'TranscriptPosEnd',\n",
       " 'TranscriptModBase',\n",
       " 'GenomicModBase',\n",
       " 'TranscriptLength',\n",
       " 'DistFromAUG',\n",
       " 'DistFromSTOP',\n",
       " 'DistFromExonStart',\n",
       " 'DistFromExonEnd',\n",
       " 'fit_c',\n",
       " 'fit_s',\n",
       " 'fit_b']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Columns that are always the same across all TSVs\n",
    "constant_cols = df.columns.tolist()[0:17]\n",
    "constant_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe53873e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clean = df.drop(columns = constant_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6d28806",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\n",
      "7KO-Cyto-BS_processed_fastqs\n"
     ]
    }
   ],
   "source": [
    "subfolder = \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\"\n",
    "folder_name = \"7KO-Cyto-BS_processed_fastqs\"\n",
    "print(subfolder, folder_name, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dd23f9c",
   "metadata": {},
   "source": [
    "### Testing make_key()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "820cabb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function definition\n",
    "def make_key(subfolder, base_key):\n",
    "    \"\"\"\n",
    "    PURPOSE:\n",
    "    Modifies names of dictionary keys based on the Rep # (detected via RegEx)\n",
    "    and Sample Type (BS, NBS) in a given subfolder name.\n",
    "    ---\n",
    "    NOTES:\n",
    "    * sorted(set(rep_matches)): Removes duplicate reps, sorts in ascending order\n",
    "    * for rep in rep_list: Adds replicate prefix to dict key names\n",
    "    * for sample in ['BS', 'NBS']: Adds sample type suffix to dict key names\n",
    "    \"\"\"\n",
    "    rep_matches = re.findall(r\"Rep\\d+\", str(subfolder))\n",
    "    rep_list = sorted(set(rep_matches))\n",
    "    \n",
    "    for rep in rep_list:\n",
    "        if f\"-{rep}-\" in str(subfolder):\n",
    "            prefix = rep + \"_\"\n",
    "            break\n",
    "    \n",
    "    for sample in [\"BS\", \"NBS\"]:\n",
    "        if f\"-{sample}_\" in str(subfolder):\n",
    "            suffix = \"_\" + sample\n",
    "            break\n",
    "    \n",
    "    return prefix + base_key + suffix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "072de3ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING REP_LIST\n",
      "Expected: ['Rep1', 'Rep2', 'Rep3'], Actual:  ['Rep1', 'Rep2', 'Rep3']\n"
     ]
    }
   ],
   "source": [
    "## Showcasing how rep_list is supposed to behave\n",
    "## Reps should be listed in integer order\n",
    "folder_list = [\"KEH-Rep3-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\"]\n",
    "rep_matches = re.findall(r\"Rep\\d+\", str(folder_list))\n",
    "rep_list = sorted(set(rep_matches))\n",
    "\n",
    "print(\"NOW TESTING REP_LIST\")\n",
    "print(\"Expected: ['Rep1', 'Rep2', 'Rep3'], Actual: \", rep_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "41a0d4a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING KEY VARIABLE\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'A': 'Rep1_A_BS',\n",
       " 'C': 'Rep1_C_BS',\n",
       " 'G': 'Rep1_G_BS',\n",
       " 'T': 'Rep1_T_BS',\n",
       " 'Deletions': 'Rep1_Deletions_BS',\n",
       " 'DeletionRate': 'Rep1_DeletionRate_BS',\n",
       " 'RealRate': 'Rep1_RealRate_BS'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key = {base_key: make_key(subfolder, base_key) for base_key \n",
    "       in [\"A\", \"C\", \"G\", \"T\", \"Deletions\", \"DeletionRate\", \"RealRate\"]}\n",
    "\n",
    "print(\"NOW TESTING KEY VARIABLE\")\n",
    "key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e69ee094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep1_DeletionRate_BS</th>\n",
       "      <th>Rep1_RealRate_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  Rep1_Deletions_BS  \\\n",
       "0        0.0        0.0        0.0        0.0                2.0   \n",
       "1        0.0        0.0        0.0        0.0                2.0   \n",
       "2        0.0        0.0        0.0        0.0                2.0   \n",
       "3        0.0        0.0        0.0        0.0                2.0   \n",
       "4        0.0        0.0        0.0        0.0                2.0   \n",
       "\n",
       "   Rep1_DeletionRate_BS  Rep1_RealRate_BS  \n",
       "0                   1.0          1.058417  \n",
       "1                   1.0          1.058417  \n",
       "2                   1.0          1.058417  \n",
       "3                   1.0          1.058417  \n",
       "4                   1.0          1.058417  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Dynamically rename columns in df_clean\n",
    "df_final = df_clean.rename(columns = {\"A\": key[\"A\"], \n",
    "                                      \"C\": key[\"C\"], \n",
    "                                      \"G\": key[\"G\"], \n",
    "                                      \"T\": key[\"T\"], \n",
    "                                      \"Deletions\": key[\"Deletions\"], \n",
    "                                      \"DeletionRate\": key[\"DeletionRate\"], \n",
    "                                      \"RealRate\": key[\"RealRate\"]})\n",
    "\n",
    "df_final.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac3f6f69",
   "metadata": {},
   "source": [
    "### Testing DeletionRate filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ccf9783b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected: Rep1_DeletionRate_BS, Actual: Rep1_DeletionRate_BS\n",
      "Size of original df: 79249\n",
      "Size of filtered df: 63189\n"
     ]
    }
   ],
   "source": [
    "## Filter by DeletionRate based on keywords in \n",
    "## folder_name and DeletionRate column (accessed via key)\n",
    "\"\"\"\n",
    "WT:\n",
    "* BS files must have DeletionRate values of >= 0.8\n",
    "* NBS files must have DeletionRate values of <= 0.1\n",
    "---\n",
    "Mutation (PUS7KO):\n",
    "* BS files must have DeletionRate values of <= 0.1\n",
    "\"\"\"\n",
    "dr_pattern = key[\"DeletionRate\"]\n",
    "print(\"Expected: Rep1_DeletionRate_BS, Actual:\", dr_pattern)\n",
    "\n",
    "if re.match(fr\"(WT|7KO).*\", str(folder_name)):\n",
    "    if re.match(fr\"WT.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            df_final = df_final[df_final[dr_pattern].ge(0.8)]\n",
    "        else: \n",
    "            df_final = df_final[df_final[dr_pattern].le(0.1)]\n",
    "\n",
    "    if re.match(fr\"7KO.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            df_final = df_final[df_final[dr_pattern].le(0.1)]\n",
    "\n",
    "print(f\"Size of original df: {df_clean.shape[0]}\",\n",
    "      f\"Size of filtered df: {df_final.shape[0]}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfa21cc4",
   "metadata": {},
   "source": [
    "### Grouping outputs by UNUAR motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "91e81c22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['UGUAA', 'UUUAA', 'UUUAG', 'UAUAG', 'UAUAA', 'UCUAG', 'UCUAA',\n",
       "       'UGUAG'], dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Find counts for each UNUAR motif\n",
    "df[\"Motif\"].unique()\n",
    "# df[\"Motif\"].groupby()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
