{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "626ab5ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import traceback\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "## Disable .loc indexing warning\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbf26e5a",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e45fa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_c</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>DeletionRate</th>\n",
       "      <th>RealRate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_021254.4</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1102</td>\n",
       "      <td>1107</td>\n",
       "      <td>1104</td>\n",
       "      <td>32601765</td>\n",
       "      <td>3516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_001350335.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>2193</td>\n",
       "      <td>2198</td>\n",
       "      <td>2195</td>\n",
       "      <td>32601765</td>\n",
       "      <td>2294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TranscriptID  Motif Region        Chrom Strand  TranscriptPosStart  \\\n",
       "0     rna-NM_021254.4  UGUAA   3UTR  NC_000021.9      -                1102   \n",
       "1  rna-NM_001350335.2  UGUAA   3UTR  NC_000021.9      -                2193   \n",
       "2  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "3  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "4  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "\n",
       "   TranscriptPosEnd  TranscriptModBase  GenomicModBase  TranscriptLength  ...  \\\n",
       "0              1107               1104        32601765              3516  ...   \n",
       "1              2198               2195        32601765              2294  ...   \n",
       "2              1405               1402        32601765              1501  ...   \n",
       "3              1405               1402        32601765              1501  ...   \n",
       "4              1405               1402        32601765              1501  ...   \n",
       "\n",
       "      fit_c     fit_s  fit_b    A    C    G    T  Deletions  DeletionRate  \\\n",
       "0  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "1  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "2  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "3  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "4  0.944807  0.779573  0.001  0.0  0.0  0.0  0.0        2.0           1.0   \n",
       "\n",
       "   RealRate  \n",
       "0  1.058417  \n",
       "1  1.058417  \n",
       "2  1.058417  \n",
       "3  1.058417  \n",
       "4  1.058417  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load in file\n",
    "df = pd.read_csv(\"C:/Users/Sonia Ling/Desktop/calculate_dr/KEH-Rep2-7KO-HEK293T-Cyto-BS.sorted.tsv\", sep = \"\\t\")\n",
    "\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c6689f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TranscriptID',\n",
       " 'Motif',\n",
       " 'Region',\n",
       " 'Chrom',\n",
       " 'Strand',\n",
       " 'TranscriptPosStart',\n",
       " 'TranscriptPosEnd',\n",
       " 'TranscriptModBase',\n",
       " 'GenomicModBase',\n",
       " 'TranscriptLength',\n",
       " 'DistFromAUG',\n",
       " 'DistFromSTOP',\n",
       " 'DistFromExonStart',\n",
       " 'DistFromExonEnd',\n",
       " 'fit_c',\n",
       " 'fit_s',\n",
       " 'fit_b']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Columns that are always the same across all TSVs\n",
    "constant_cols = df.columns.tolist()[0:17]\n",
    "constant_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fe53873e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clean = df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a6d28806",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\n",
      "7KO-Cyto-BS_processed_fastqs\n"
     ]
    }
   ],
   "source": [
    "subfolder = \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\"\n",
    "folder_name = \"7KO-Cyto-BS_processed_fastqs\"\n",
    "print(subfolder, folder_name, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dd23f9c",
   "metadata": {},
   "source": [
    "### Testing make_key()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "820cabb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function definition\n",
    "def make_key(subfolder, base_key):\n",
    "    \"\"\"\n",
    "    PURPOSE:\n",
    "    Modifies names of dictionary keys based on the Rep # (detected via RegEx)\n",
    "    and Sample Type (BS, NBS) in a given subfolder name.\n",
    "    ---\n",
    "    NOTES:\n",
    "    * sorted(set(rep_matches)): Removes duplicate reps, sorts in ascending order\n",
    "    * for rep in rep_list: Adds replicate prefix to dict key names\n",
    "    * for sample in ['BS', 'NBS']: Adds sample type suffix to dict key names\n",
    "    \"\"\"\n",
    "    rep_matches = re.findall(r\"Rep\\d+\", str(subfolder))\n",
    "    rep_list = sorted(set(rep_matches))\n",
    "    \n",
    "    for rep in rep_list:\n",
    "        if f\"-{rep}-\" in str(subfolder):\n",
    "            prefix = rep + \"_\"\n",
    "            break\n",
    "    \n",
    "    for sample in [\"BS\", \"NBS\"]:\n",
    "        if f\"-{sample}_\" in str(subfolder):\n",
    "            suffix = \"_\" + sample\n",
    "            break\n",
    "    \n",
    "    return prefix + base_key + suffix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "072de3ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING REP_LIST\n",
      "Expected: ['Rep1', 'Rep2', 'Rep3'], Actual:  ['Rep1', 'Rep2', 'Rep3']\n",
      "NOW TESTING REP_LIST\n",
      "Expected: ['Rep1', 'Rep5', 'Rep10'], Actual:  ['Rep1', 'Rep5', 'Rep10']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "PURPOSE:\n",
    "* Showcases how rep_list is supposed to bheave\n",
    "* Reps should be listed in integer order\n",
    "\"\"\"\n",
    "\n",
    "## Test 1: Rep1, Rep2, Rep3\n",
    "folder_list = [\"KEH-Rep3-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\"]\n",
    "rep_matches = re.findall(r\"Rep\\d+\", str(folder_list))\n",
    "rep_list = sorted(set(rep_matches), key = lambda x: int(x[3:]))\n",
    "\n",
    "print(\"NOW TESTING REP_LIST\")\n",
    "print(\"Expected: ['Rep1', 'Rep2', 'Rep3'], Actual: \", rep_list)\n",
    "\n",
    "\n",
    "## Test 2: Rep1, Rep5, Rep10\n",
    "folder_list = [\"KEH-Rep10-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep5-7KO-HEK293T-Cyto-BS_star_realigned\"]\n",
    "rep_matches = re.findall(r\"Rep\\d+\", str(folder_list))\n",
    "rep_list = sorted(set(rep_matches), key = lambda x: int(x[3:]))\n",
    "\n",
    "print(\"NOW TESTING REP_LIST\")\n",
    "print(\"Expected: ['Rep1', 'Rep5', 'Rep10'], Actual: \", rep_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "41a0d4a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING KEY VARIABLE\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'A': 'Rep1_A_BS',\n",
       " 'C': 'Rep1_C_BS',\n",
       " 'G': 'Rep1_G_BS',\n",
       " 'T': 'Rep1_T_BS',\n",
       " 'Deletions': 'Rep1_Deletions_BS',\n",
       " 'DeletionRate': 'Rep1_DeletionRate_BS',\n",
       " 'RealRate': 'Rep1_RealRate_BS'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key = {base_key: make_key(subfolder, base_key) for base_key \n",
    "       in [\"A\", \"C\", \"G\", \"T\", \"Deletions\", \"DeletionRate\", \"RealRate\"]}\n",
    "\n",
    "print(\"NOW TESTING KEY VARIABLE\")\n",
    "key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e69ee094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_c</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep1_DeletionRate_BS</th>\n",
       "      <th>Rep1_RealRate_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_021254.4</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1102</td>\n",
       "      <td>1107</td>\n",
       "      <td>1104</td>\n",
       "      <td>32601765</td>\n",
       "      <td>3516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_001350335.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>2193</td>\n",
       "      <td>2198</td>\n",
       "      <td>2195</td>\n",
       "      <td>32601765</td>\n",
       "      <td>2294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_001350337.2</td>\n",
       "      <td>UGUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>-</td>\n",
       "      <td>1400</td>\n",
       "      <td>1405</td>\n",
       "      <td>1402</td>\n",
       "      <td>32601765</td>\n",
       "      <td>1501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944807</td>\n",
       "      <td>0.779573</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TranscriptID  Motif Region        Chrom Strand  TranscriptPosStart  \\\n",
       "0     rna-NM_021254.4  UGUAA   3UTR  NC_000021.9      -                1102   \n",
       "1  rna-NM_001350335.2  UGUAA   3UTR  NC_000021.9      -                2193   \n",
       "2  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "3  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "4  rna-NM_001350337.2  UGUAA   3UTR  NC_000021.9      -                1400   \n",
       "\n",
       "   TranscriptPosEnd  TranscriptModBase  GenomicModBase  TranscriptLength  ...  \\\n",
       "0              1107               1104        32601765              3516  ...   \n",
       "1              2198               2195        32601765              2294  ...   \n",
       "2              1405               1402        32601765              1501  ...   \n",
       "3              1405               1402        32601765              1501  ...   \n",
       "4              1405               1402        32601765              1501  ...   \n",
       "\n",
       "      fit_c     fit_s  fit_b  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  0.944807  0.779573  0.001        0.0        0.0        0.0        0.0   \n",
       "1  0.944807  0.779573  0.001        0.0        0.0        0.0        0.0   \n",
       "2  0.944807  0.779573  0.001        0.0        0.0        0.0        0.0   \n",
       "3  0.944807  0.779573  0.001        0.0        0.0        0.0        0.0   \n",
       "4  0.944807  0.779573  0.001        0.0        0.0        0.0        0.0   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep1_DeletionRate_BS  Rep1_RealRate_BS  \n",
       "0                2.0                   1.0          1.058417  \n",
       "1                2.0                   1.0          1.058417  \n",
       "2                2.0                   1.0          1.058417  \n",
       "3                2.0                   1.0          1.058417  \n",
       "4                2.0                   1.0          1.058417  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Dynamically rename columns in df_clean\n",
    "df_final = df_clean.rename(columns = {\"A\": key[\"A\"], \n",
    "                                      \"C\": key[\"C\"], \n",
    "                                      \"G\": key[\"G\"], \n",
    "                                      \"T\": key[\"T\"], \n",
    "                                      \"Deletions\": key[\"Deletions\"], \n",
    "                                      \"DeletionRate\": key[\"DeletionRate\"], \n",
    "                                      \"RealRate\": key[\"RealRate\"]})\n",
    "\n",
    "df_final.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c52c92",
   "metadata": {},
   "source": [
    "### Testing get_sample_group()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b0cb1263",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function definition\n",
    "def get_sample_group(folder_name):\n",
    "    \"\"\"\n",
    "    PURPOSE:\n",
    "    Given input folder names, extract the group name\n",
    "    by returning the first capture group in RegEx.\n",
    "    ---\n",
    "    EXAMPLE: \n",
    "    '7KO-Cyto-BS_processed_fastqs' -> '7KO-Cyto'\n",
    "    \"\"\"\n",
    "    match = re.match(r\"(.+)-(?:BS|NBS)_processed_fastqs\", folder_name)\n",
    "    return match.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "38256626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FOLDER NAME: 7KO-Cyto-BS_processed_fastqs\n",
      "Expected: 7KO-Cyto, Actual: 7KO-Cyto\n",
      "Expected: 7KO-Cyto, Actual: 7KO-Cyto\n",
      "Expected: WT-Nuc, Actual: WT-Nuc\n",
      "Expected: WT-Nuc, Actual: WT-Nuc\n"
     ]
    }
   ],
   "source": [
    "print(\"FOLDER NAME:\", folder_name)\n",
    "print(\"Expected: 7KO-Cyto, Actual:\", get_sample_group(folder_name))\n",
    "print(\"Expected: 7KO-Cyto, Actual:\", get_sample_group(\"7KO-Cyto-NBS_processed_fastqs\"))\n",
    "print(\"Expected: WT-Nuc, Actual:\", get_sample_group(\"WT-Nuc-BS_processed_fastqs\"))\n",
    "print(\"Expected: WT-Nuc, Actual:\", get_sample_group(\"WT-Nuc-NBS_processed_fastqs\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4735b30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ac3f6f69",
   "metadata": {},
   "source": [
    "### Testing filtering & cutoffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ccf9783b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected: Rep1_DeletionRate_BS, Actual: Rep1_DeletionRate_BS\n",
      "Expected: Rep1_RealRate_BS, Actual: Rep1_RealRate_BS\n"
     ]
    }
   ],
   "source": [
    "## Filter by DeletionRate based on keywords in \n",
    "## folder_name and DeletionRate column (accessed via key)\n",
    "\"\"\"\n",
    "WT:\n",
    "* BS files must have DeletionRate values of >= 0.8\n",
    "* NBS files must have DeletionRate values of <= 0.1\n",
    "---\n",
    "Mutation (PUS7KO):\n",
    "* BS files must have DeletionRate values of <= 0.1\n",
    "\"\"\"\n",
    "dr_pattern = key[\"DeletionRate\"]\n",
    "print(\"Expected: Rep1_DeletionRate_BS, Actual:\", dr_pattern)\n",
    "\n",
    "rr_pattern = key[\"RealRate\"]\n",
    "print(\"Expected: Rep1_RealRate_BS, Actual:\", rr_pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a16701da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of original df: 79249\n",
      "Size of kept_rr: 70129\n",
      "\n",
      "Size of original df: 79249\n",
      "Size of dropped_rr: 9120\n",
      "\n",
      "Expected size of dropped_rr: 9120\n",
      "Actual: 9120\n"
     ]
    }
   ],
   "source": [
    "## Testing RealRate filtering\n",
    "kept_rr = df_final[df_final[rr_pattern].ge(0.3)]\n",
    "print(f\"Size of original df: {len(df_clean)}\",\n",
    "      f\"Size of kept_rr: {len(kept_rr)}\", sep = \"\\n\")\n",
    "\n",
    "dropped_rr = df_final[df_final[rr_pattern].lt(0.3)]\n",
    "print(f\"\\nSize of original df: {len(df_clean)}\",\n",
    "      f\"Size of dropped_rr: {len(dropped_rr)}\", sep = \"\\n\")\n",
    "\n",
    "print(f\"\\nExpected size of dropped_rr: {len(df_clean) - len(kept_rr)}\",\n",
    "      f\"Actual: {len(dropped_rr)}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b1766b2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of kept_rr: 70129\n",
      "Size of kept_cov: 63439\n"
     ]
    }
   ],
   "source": [
    "## Testing coverage filtering\n",
    "coverage_list = [col for col in kept_rr.columns \n",
    "                 if re.search(\"(A|C|G|T|Deletions)_.*\", col)]\n",
    "kept_rr[\"TotalCoverage\"] = kept_rr[coverage_list].sum(axis = 1)\n",
    "kept_cov = kept_rr[kept_rr[\"TotalCoverage\"].ge(20)]\n",
    "\n",
    "print(f\"Size of kept_rr: {len(kept_rr)}\",\n",
    "      f\"Size of kept_cov: {len(kept_cov)}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8fe10a35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of original df: 79249\n",
      "Size of kept_rr: 70129\n",
      "Size of kept_cov: 63439\n",
      "Size of final filtered df: 54052\n",
      "\n",
      "Total rows removed: 25197\n",
      "The dataset shrunk by approximately 32% after filtering\n"
     ]
    }
   ],
   "source": [
    "## Testing DeletionRate filtering\n",
    "if re.match(fr\"(WT|7KO).*\", str(folder_name)):\n",
    "    if re.match(fr\"WT.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            df_final = kept_cov[kept_cov[dr_pattern].ge(0.8)]\n",
    "        else: \n",
    "            df_final = kept_cov[kept_cov[dr_pattern].le(0.1)]\n",
    "\n",
    "    if re.match(fr\"7KO.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            ## THIS LINE SHOULD EXECUTE\n",
    "            ## folder_name = 7KO-Cyto-BS_processed_fastqs\n",
    "            df_final = kept_cov[kept_cov[dr_pattern].le(0.1)]\n",
    "\n",
    "    df_final = df_final.sort_values(by = dr_pattern, ascending = False)\n",
    "\n",
    "print(f\"Size of original df: {len(df_clean)}\",\n",
    "      f\"Size of kept_rr: {len(kept_rr)}\",\n",
    "      f\"Size of kept_cov: {len(kept_cov)}\",\n",
    "      f\"Size of final filtered df: {len(df_final)}\", sep = \"\\n\")\n",
    "\n",
    "rows_removed = (len(df_clean) - len(df_final))\n",
    "percent_change = round((rows_removed/len(df_clean)) * 100)\n",
    "print(f\"\\nTotal rows removed: {rows_removed}\",\n",
    "      f\"The dataset shrunk by approximately {percent_change}% after filtering\",\n",
    "      sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfa21cc4",
   "metadata": {},
   "source": [
    "### Investigating results of df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2abbe280",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep1_DeletionRate_BS</th>\n",
       "      <th>Rep1_RealRate_BS</th>\n",
       "      <th>TotalCoverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16319</th>\n",
       "      <td>rna-XM_005245453.2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3857</td>\n",
       "      <td>3862</td>\n",
       "      <td>3859</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16318</th>\n",
       "      <td>rna-XM_005245453.2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3857</td>\n",
       "      <td>3862</td>\n",
       "      <td>3859</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16320</th>\n",
       "      <td>rna-NM_022731.5</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3860</td>\n",
       "      <td>3865</td>\n",
       "      <td>3862</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16321</th>\n",
       "      <td>rna-NM_022731.5</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3860</td>\n",
       "      <td>3865</td>\n",
       "      <td>3862</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15814</th>\n",
       "      <td>rna-NM_012458.4</td>\n",
       "      <td>UAUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>-</td>\n",
       "      <td>1250</td>\n",
       "      <td>1255</td>\n",
       "      <td>1252</td>\n",
       "      <td>2426036</td>\n",
       "      <td>1664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.966038</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.735916</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16330</th>\n",
       "      <td>rna-NM_001244926.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1688</td>\n",
       "      <td>1693</td>\n",
       "      <td>1690</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2897</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16331</th>\n",
       "      <td>rna-NM_001244926.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1688</td>\n",
       "      <td>1693</td>\n",
       "      <td>1690</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2897</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16338</th>\n",
       "      <td>rna-NM_004697.5</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1691</td>\n",
       "      <td>1696</td>\n",
       "      <td>1693</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16337</th>\n",
       "      <td>rna-NM_004697.5</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1691</td>\n",
       "      <td>1696</td>\n",
       "      <td>1693</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16336</th>\n",
       "      <td>rna-NM_004697.5</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1691</td>\n",
       "      <td>1696</td>\n",
       "      <td>1693</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TranscriptID  Motif Region         Chrom Strand  \\\n",
       "16319  rna-XM_005245453.2  UUUAG   3UTR  NC_000001.11      -   \n",
       "16318  rna-XM_005245453.2  UUUAG   3UTR  NC_000001.11      -   \n",
       "16320     rna-NM_022731.5  UUUAG   3UTR  NC_000001.11      -   \n",
       "16321     rna-NM_022731.5  UUUAG   3UTR  NC_000001.11      -   \n",
       "15814     rna-NM_012458.4  UAUAG   3UTR  NC_000019.10      -   \n",
       "16330  rna-NM_001244926.2  UCUAA   3UTR  NC_000009.12      +   \n",
       "16331  rna-NM_001244926.2  UCUAA   3UTR  NC_000009.12      +   \n",
       "16338     rna-NM_004697.5  UCUAA   3UTR  NC_000009.12      +   \n",
       "16337     rna-NM_004697.5  UCUAA   3UTR  NC_000009.12      +   \n",
       "16336     rna-NM_004697.5  UCUAA   3UTR  NC_000009.12      +   \n",
       "\n",
       "       TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  \\\n",
       "16319                3857              3862               3859   \n",
       "16318                3857              3862               3859   \n",
       "16320                3860              3865               3862   \n",
       "16321                3860              3865               3862   \n",
       "15814                1250              1255               1252   \n",
       "16330                1688              1693               1690   \n",
       "16331                1688              1693               1690   \n",
       "16338                1691              1696               1693   \n",
       "16337                1691              1696               1693   \n",
       "16336                1691              1696               1693   \n",
       "\n",
       "       GenomicModBase  TranscriptLength  ...     fit_s  fit_b  Rep1_A_BS  \\\n",
       "16319       205715358              6396  ...  0.901845  0.001      249.0   \n",
       "16318       205715358              6396  ...  0.901845  0.001      249.0   \n",
       "16320       205715358              6399  ...  0.901845  0.001      249.0   \n",
       "16321       205715358              6399  ...  0.901845  0.001      249.0   \n",
       "15814         2426036              1664  ...  0.966038  0.001        8.0   \n",
       "16330       113291699              2897  ...  0.890419  0.001        0.0   \n",
       "16331       113291699              2897  ...  0.890419  0.001        0.0   \n",
       "16338       113291699              2900  ...  0.890419  0.001        0.0   \n",
       "16337       113291699              2900  ...  0.890419  0.001        0.0   \n",
       "16336       113291699              2900  ...  0.890419  0.001        0.0   \n",
       "\n",
       "       Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  Rep1_Deletions_BS  \\\n",
       "16319        0.0        0.0        0.0               24.0   \n",
       "16318        0.0        0.0        0.0               24.0   \n",
       "16320        0.0        0.0        0.0               24.0   \n",
       "16321        0.0        0.0        0.0               24.0   \n",
       "15814       15.0        0.0        0.0                2.0   \n",
       "16330        0.0        0.0       25.0                2.0   \n",
       "16331        0.0        0.0       25.0                2.0   \n",
       "16338        0.0        0.0       25.0                2.0   \n",
       "16337        0.0        0.0       25.0                2.0   \n",
       "16336        0.0        0.0       25.0                2.0   \n",
       "\n",
       "       Rep1_DeletionRate_BS  Rep1_RealRate_BS  TotalCoverage  \n",
       "16319              0.087912          0.506265          273.0  \n",
       "16318              0.087912          0.506265          273.0  \n",
       "16320              0.087912          0.506265          273.0  \n",
       "16321              0.087912          0.506265          273.0  \n",
       "15814              0.080000          0.735916           25.0  \n",
       "16330              0.074074          0.426971           27.0  \n",
       "16331              0.074074          0.426971           27.0  \n",
       "16338              0.074074          0.426971           27.0  \n",
       "16337              0.074074          0.426971           27.0  \n",
       "16336              0.074074          0.426971           27.0  \n",
       "\n",
       "[10 rows x 25 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9ad79f11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The only UNUAR motif missing in the final dataframe is UGUAA.\n"
     ]
    }
   ],
   "source": [
    "original_motifs = df_clean[\"Motif\"].unique()\n",
    "final_motifs = df_final[\"Motif\"].unique()\n",
    "set_diff = list(set(original_motifs) - set(final_motifs))\n",
    "\n",
    "print(f\"The only UNUAR motif missing in the final dataframe is {set_diff[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5c80bdcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## For priority_filtered\n",
    "# df_final = df.drop(columns = constant_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e81c22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['UGUAA', 'UUUAA', 'UUUAG', 'UAUAG', 'UAUAA', 'UCUAG', 'UCUAA',\n",
       "       'UGUAG'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Find counts for each UNUAR motif\n",
    "# df[\"Motif\"].groupby()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
