{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ebdf09f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import traceback\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from scipy.stats import fisher_exact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d48ade8",
   "metadata": {},
   "source": [
    "## Class initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f88fa575",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FilterTSV:\n",
    "   def create_mask(self, df, colnames):\n",
    "      \"\"\"\n",
    "      NOTES:\n",
    "      * Select columns that contain \"Deletions\" and put them in a list\n",
    "      * Use set() to remove duplicates, since sets can only contain unique vals\n",
    "      * Pass column names in list to dataframe to create a mask that drops rows\n",
    "        where Deletions == 0 and there are nulls\n",
    "      \"\"\"\n",
    "      del_list = list(set([col for col in colnames if re.search(r\"Deletions\", col)]))\n",
    "      mask = ~(df[del_list] == 0).any(axis = 1) & (df.notna().all(axis = 1))\n",
    "      return mask\n",
    "\n",
    "   def merged_output(self, df_merged, rep_list, pattern_dict):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      1. Takes all columns from merged df and organizes them by BS/NBS type \n",
    "      2. Sums up corresponding bases/deletions & creates 4 new columns per replicate\n",
    "      3. Selects the new columns\n",
    "         * Reshapes each row into 2x2 matrix\n",
    "         * Runs Fisher's Exact Test\n",
    "         * Appends p-val column\n",
    "      \"\"\"\n",
    "      try:\n",
    "         for rep in rep_list:\n",
    "            ## Define names of summed BS/NBS columns\n",
    "            new_cols = [f\"{rep}_TotalBases_BS\", \n",
    "                        f\"{rep}_TotalBases_NBS\"]\n",
    "            \n",
    "            ## Define generic entries for 2x2 contingency table\n",
    "            fisher_cols = [f\"{rep}_TotalBases_BS\", \n",
    "                           f\"{rep}_Deletions_BS\", \n",
    "                           f\"{rep}_TotalBases_NBS\", \n",
    "                           f\"{rep}_Deletions_NBS\"]\n",
    "\n",
    "            ## Calculate p-values\n",
    "            \"\"\"\n",
    "            PART I: For each replicate, find total bases for BS/NBS\n",
    "            * Zips pattern_dict and new_cols together\n",
    "              -> Reminder: pattern_dict = {[List of BS colnames], [List of NBS colnames]}\n",
    "                           new_cols = [\"TotalBases_BS\", \"TotalBases_NBS\"]\n",
    "            * Sum of all entries in 1st list of pattern_dict \n",
    "              -> Stored under 1st colname in new_cols\n",
    "            * Sum of all entries in 2nd list of pattern_dict \n",
    "              -> Stored under 2nd colname in new_cols \n",
    "            \"\"\"\n",
    "            for col, key in zip(new_cols, pattern_dict):\n",
    "               if col not in df_merged.columns:\n",
    "                  df_merged[col] = df_merged[pattern_dict[key]].sum(axis=1)\n",
    "\n",
    "            \"\"\"\n",
    "            PART II: Run Fisher's Exact Test using 2x2 table\n",
    "            * For each row in df_merged:\n",
    "              -> Select the specified columns from fisher_cols\n",
    "              -> Reshape the 4 columns into separate 3D arrays of size 2x2\n",
    "                 (these will be our 2x2 tables)\n",
    "              -> Use these arrays for numpy batch processing\n",
    "            * Run Fisher's Exact Test (scipy) on each table, then select second result\n",
    "              of test AKA the p-val using 'fisher_exact(table)[1]'\n",
    "            \"\"\"\n",
    "            if set(fisher_cols).issubset(df_merged.columns):\n",
    "               df_merged = df_merged.dropna(subset = fisher_cols)\n",
    "               arr = df_merged[fisher_cols].values.reshape(-1, 2, 2) \n",
    "               pvals = [fisher_exact(table)[1] for table in arr]\n",
    "               df_merged[f\"{rep}_Pvalue\"] = pvals\n",
    "         return df_merged\n",
    "      except Exception as e:\n",
    "         print(f\"Failed to calculate p-value for {rep}: {e}\")\n",
    "         traceback.print_exc()\n",
    "         raise\n",
    "\n",
    "   def filter_means(self, df_filtered, colname, cols):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      * Use to filter by average (Cutoffs #4-6)\n",
    "      \"\"\"\n",
    "      ## Calculate average and standard deviation\n",
    "      df_filtered[colname] = df_filtered[cols].mean(axis = 1)      \n",
    "      std_colname = colname.replace(\"Avg\", \"Std\")\n",
    "      df_filtered[std_colname] = df_filtered[cols].std(axis = 1)\n",
    "\n",
    "      ## Sort by descending DeletionRate\n",
    "      df_filtered = df_filtered.sort_values(by = colname, \n",
    "                                            ascending = False)\n",
    "\n",
    "      ## If BS, apply filters to average columns\n",
    "      if \"_BS\" in colname:\n",
    "         if \"DeletionCt\" in colname:\n",
    "            df_filtered[colname] = df_filtered[colname].ge(5)\n",
    "         elif \"DeletionRate\" in colname:\n",
    "            df_filtered[colname] = df_filtered[colname].ge(0.02)\n",
    "\n",
    "      return df_filtered\n",
    "\n",
    "   def filtered_output(self, df_merged, rep_list):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      a) Adds cutoffs from BID-Pipe protocol:\n",
    "         1. Pvalue across all replicates <= 0.0001\n",
    "         2. RealRate across all replicates >= 0.3\n",
    "         3. Total sequencing coverage for each BS and NBS replicate >= 20\n",
    "         4. Average Deletions across all BS replicates >= 5\n",
    "         5. Average DeletionRate across all BS replicates >= 0.02\n",
    "         6. Average DeletionRate is 2x higher in BS compared to NBS\n",
    "      b) Saves filtered and discarded rows in separate dataframes\n",
    "      \"\"\"\n",
    "      try:\n",
    "         ## Cutoff 1: Pvalue\n",
    "         pval_list = [col for col in df_merged.columns \n",
    "                      if re.search(r\"Pvalue$\", col)]\n",
    "         cutoff1 = df_merged[pval_list].le(0.0001).all(axis=1)\n",
    "         df_filtered = df_merged.loc[cutoff1]\n",
    "\n",
    "         ## Cutoff 2: RealRate\n",
    "         realrate_list = [col for col in df_filtered.columns \n",
    "                          if re.search(r\"RealRate\", col)]\n",
    "         cutoff2 = df_filtered[realrate_list].ge(0.3).all(axis=1)\n",
    "         df_filtered = df_filtered.loc[cutoff2]\n",
    "\n",
    "         ## Cutoff 3: Total sequencing coverage\n",
    "         for rep in rep_list:\n",
    "            for sample in [\"BS\", \"NBS\"]:\n",
    "               coverage_list = [col for col in df_filtered.columns if \n",
    "                                re.match(fr\"{rep}_(TotalBases|Deletions)_{sample}\", col)]\n",
    "               total_sum = df_filtered[coverage_list].sum(axis = 1)\n",
    "               cutoff3 = total_sum.ge(20)\n",
    "               df_filtered = df_filtered.loc[cutoff3]\n",
    "\n",
    "         ## Cutoff 4: Average Deletions (BS)\n",
    "         avg_del_bs = \"AvgDeletionCt_BS\"\n",
    "         del_col_bs = [col for col in df_filtered.columns \n",
    "                       if re.search(r\"_Deletions_BS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_del_bs, del_col_bs)\n",
    "\n",
    "         ## Cutoff 5: Average DeletionRate (BS)\n",
    "         avg_dr_bs = \"AvgDeletionRate_BS\"\n",
    "         dr_col_bs = [col for col in df_filtered.columns \n",
    "                      if re.search(r\"_DeletionRate_BS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_dr_bs, dr_col_bs)\n",
    "\n",
    "         ## Cutoff 6: Average DeletionRate is 2x higher in BS compared to NBS\n",
    "         avg_dr_nbs = \"AvgDeletionRate_NBS\"\n",
    "         dr_col_nbs = [col for col in df_filtered.columns \n",
    "                       if re.search(r\"_DeletionRate_NBS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_dr_nbs, dr_col_nbs)\n",
    "         \n",
    "         cutoff6 = df_filtered[avg_dr_bs] >= 2 * df_filtered[avg_dr_nbs]\n",
    "         df_filtered = df_filtered[cutoff6]\n",
    "\n",
    "         print(\"Successfully applied cutoffs.\")\n",
    "\n",
    "         return df_filtered\n",
    "      \n",
    "      except Exception as e:\n",
    "         print(f\"Failed to apply cutoffs from BID-Pipe protocol: {e}\")\n",
    "         traceback.print_exc()\n",
    "         raise\n",
    "\n",
    "filtertsv = FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7213f1ce",
   "metadata": {},
   "source": [
    "## Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1a82247a",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_path = Path.cwd()\n",
    "input_dir = current_path/\"calculations\"\n",
    "filtertsv = FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba0541a",
   "metadata": {},
   "source": [
    "## Testing helper functions in FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d2dc3db",
   "metadata": {},
   "source": [
    "### Testing create_mask() with random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a2ad7424",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  \n",
       "0                  2  \n",
       "1                  5  \n",
       "2                  3  \n",
       "3                  0  \n",
       "4                  0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Testing del_list variable\n",
    "\n",
    "random_list = [{\"Chrom\": \"NW233434\", \n",
    "                \"GenomicModBase\": 231, \n",
    "                \"Rep1_A_BS\": 4, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 2}, \n",
    "\n",
    "                {\"Chrom\": \"NW233438\", \n",
    "                \"GenomicModBase\": 231, \n",
    "                \"Rep1_A_BS\": 4, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 5},\n",
    "\n",
    "                {\"Chrom\": \"NW233435\", \n",
    "                 \"GenomicModBase\": 331, \n",
    "                 \"Rep1_A_BS\": None, \n",
    "                 \"Rep1_C_BS\": None, \n",
    "                 \"Rep1_G_BS\": None, \n",
    "                 \"Rep1_T_BS\": None, \n",
    "                 \"Rep1_Deletions_BS\": 3},\n",
    "\n",
    "                {\"Chrom\": \"NW233433\", \n",
    "                \"GenomicModBase\": 230, \n",
    "                \"Rep1_A_BS\": 3, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 0},\n",
    "                \n",
    "                {\"Chrom\": \"NW233432\", \n",
    "                \"GenomicModBase\": 229, \n",
    "                \"Rep1_A_BS\": 2, \n",
    "                \"Rep1_C_BS\": 6, \n",
    "                \"Rep1_G_BS\": 4, \n",
    "                \"Rep1_T_BS\": 7, \n",
    "                \"Rep1_Deletions_BS\": 0}]\n",
    "random_df = pd.DataFrame(random_list).drop_duplicates()\n",
    "\n",
    "random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3b927b1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING DEL_LIST VARIABLE (order in list doesn't matter)\n",
      "Expected: ['Rep2_Deletions_BS', 'Rep1_Deletions_BS']\n",
      "Actual: ['Rep1_Deletions_BS']\n"
     ]
    }
   ],
   "source": [
    "del_list = list(set([col for col in random_df.columns if re.search(\"Deletions\", col)]))\n",
    "\n",
    "print(\"NOW TESTING DEL_LIST VARIABLE (order in list doesn't matter)\",\n",
    "      \"Expected: ['Rep2_Deletions_BS', 'Rep1_Deletions_BS']\",\n",
    "      f\"Actual: {del_list}\", sep = \"\\n\")\n",
    "\n",
    "## Testing mask variable\n",
    "mask = ~(random_df[del_list] == 0).any(axis = 1) & (random_df.notna().all(axis = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b75b943e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING ORIGINAL RANDOM_DF\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  \n",
       "0                  2  \n",
       "1                  5  \n",
       "2                  3  \n",
       "3                  0  \n",
       "4                  0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"\\nNOW PRINTING ORIGINAL RANDOM_DF\")\n",
    "random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f4e8496",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING FILTERED RANDOM_DF\n",
      "Expected: Rows with 0 deletions or nulls should be removed\n",
      "Actual:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "\n",
       "   Rep1_Deletions_BS  \n",
       "0                  2  \n",
       "1                  5  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"\\nNOW PRINTING FILTERED RANDOM_DF\",\n",
    "      \"Expected: Rows with 0 deletions or nulls should be removed\",\n",
    "      \"Actual:\", sep = \"\\n\")\n",
    "good_df = random_df.loc[mask]\n",
    "good_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "199b08e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXPECTED: Failed to filter out DeletionRate == 0 in bad_df\n",
      "EXPECTED: Successfully filtered out DeletionRate == 0 in good_df\n"
     ]
    }
   ],
   "source": [
    "bad_mask = (random_df[random_df[del_list] != 0].all(axis = 1)) & (~random_df.isnull().any(axis = 1))\n",
    "bad_df = random_df.loc[bad_mask]\n",
    "\n",
    "if 0 in bad_df[del_list].values:\n",
    "    print(\"EXPECTED: Failed to filter out DeletionRate == 0 in bad_df\")\n",
    "else:\n",
    "    print(\"NOT EXPECTED: Successfully filtered out DeletionRate == 0 in bad_df\")\n",
    "\n",
    "if 0 in good_df[del_list].values:\n",
    "    print(\"NOT EXPECTED: Failed to filter out DeletionRate == 0 in good_df\")\n",
    "else:\n",
    "    print(\"EXPECTED: Successfully filtered out DeletionRate == 0 in good_df\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "03f29a76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected: Rows 0, 1\n",
      "Actual:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "\n",
       "   Rep1_Deletions_BS  \n",
       "0                  2  \n",
       "1                  5  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "init_mask = filtertsv.create_mask(random_df, random_df.columns)\n",
    "df_full = random_df.loc[init_mask]\n",
    "df_dropped = random_df.loc[~init_mask]\n",
    "\n",
    "print(\"Expected: Rows 0, 1\",\n",
    "      \"Actual:\", sep = \"\\n\")\n",
    "df_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "089f8dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected: Rows 2-4 (recall that Row 1 was dropped as a duplicate)\n",
      "Actual:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  \n",
       "2                  3  \n",
       "3                  0  \n",
       "4                  0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Expected: Rows 2-4 (recall that Row 1 was dropped as a duplicate)\",\n",
    "      \"Actual:\", sep = \"\\n\")\n",
    "df_dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a02ce8df",
   "metadata": {},
   "source": [
    "## Testing iterative merging"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "898fa395",
   "metadata": {},
   "source": [
    "### Using random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8444bdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>245</td>\n",
       "      <td>75</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>654</td>\n",
       "      <td>243</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SW233487</td>\n",
       "      <td>76</td>\n",
       "      <td>3</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SW233426</td>\n",
       "      <td>229</td>\n",
       "      <td>26</td>\n",
       "      <td>356.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  \\\n",
       "0  NW233434             231          7        1.0       66.0       97.0   \n",
       "1  NW233434             245         75        3.0        1.0        4.0   \n",
       "2  NW233435             654        243        NaN        NaN        NaN   \n",
       "3  SW233487              76          3       42.0        1.0        2.0   \n",
       "4  SW233426             229         26      356.0       42.0       74.0   \n",
       "\n",
       "   Rep2_Deletions_BS  \n",
       "0                 12  \n",
       "1                  7  \n",
       "2                 53  \n",
       "3                  0  \n",
       "4                  6  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Testing del_list variable\n",
    "random2_list = [{\"Chrom\": \"NW233434\", \n",
    "                \"GenomicModBase\": 231, \n",
    "                \"Rep2_A_BS\": 7, \n",
    "                \"Rep2_C_BS\": 1, \n",
    "                \"Rep2_G_BS\": 66, \n",
    "                \"Rep2_T_BS\": 97, \n",
    "                \"Rep2_Deletions_BS\": 12}, \n",
    "\n",
    "                {\"Chrom\": \"NW233434\", \n",
    "                \"GenomicModBase\": 245, \n",
    "                \"Rep2_A_BS\": 75, \n",
    "                \"Rep2_C_BS\": 3, \n",
    "                \"Rep2_G_BS\": 1, \n",
    "                \"Rep2_T_BS\": 4, \n",
    "                \"Rep2_Deletions_BS\": 7},\n",
    "\n",
    "                {\"Chrom\": \"NW233435\", \n",
    "                 \"GenomicModBase\": 654, \n",
    "                 \"Rep2_A_BS\": 243, \n",
    "                 \"Rep2_C_BS\": None, \n",
    "                 \"Rep2_G_BS\": None, \n",
    "                 \"Rep2_T_BS\": None, \n",
    "                 \"Rep2_Deletions_BS\": 53},\n",
    "\n",
    "                {\"Chrom\": \"SW233487\", \n",
    "                \"GenomicModBase\": 76, \n",
    "                \"Rep2_A_BS\": 3, \n",
    "                \"Rep2_C_BS\": 42, \n",
    "                \"Rep2_G_BS\": 1, \n",
    "                \"Rep2_T_BS\": 2, \n",
    "                \"Rep2_Deletions_BS\": 0},\n",
    "                \n",
    "                {\"Chrom\": \"SW233426\", \n",
    "                \"GenomicModBase\": 229, \n",
    "                \"Rep2_A_BS\": 26, \n",
    "                \"Rep2_C_BS\": 356, \n",
    "                \"Rep2_G_BS\": 42, \n",
    "                \"Rep2_T_BS\": 74, \n",
    "                \"Rep2_Deletions_BS\": 6}]\n",
    "random_df2 = pd.DataFrame(random2_list).drop_duplicates()\n",
    "\n",
    "random_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d7a64937",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list_ran = [random_df, random_df2]\n",
    "colnames_ran = random_df.columns.tolist()\n",
    "selected_cols_ran = colnames_ran[0:2]\n",
    "init_mask_ran = filtertsv.create_mask(random_df, colnames_ran)\n",
    "df_full_ran = random_df.loc[init_mask_ran]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9ae45635",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected df after iterative merging:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SW233426</td>\n",
       "      <td>229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233434             245        NaN        NaN        NaN        NaN   \n",
       "2  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "3  SW233426             229        NaN        NaN        NaN        NaN   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  \\\n",
       "0                2.0        7.0        1.0       66.0       97.0   \n",
       "1                NaN       75.0        3.0        1.0        4.0   \n",
       "2                5.0        NaN        NaN        NaN        NaN   \n",
       "3                NaN       26.0      356.0       42.0       74.0   \n",
       "\n",
       "   Rep2_Deletions_BS  \n",
       "0               12.0  \n",
       "1                7.0  \n",
       "2                NaN  \n",
       "3                6.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## View output from manually merged dataframe\n",
    "temp1_init_mask = filtertsv.create_mask(random_df, random_df.columns)\n",
    "temp2_init_mask = filtertsv.create_mask(random_df2, random_df2.columns)\n",
    "\n",
    "print(\"Expected df after iterative merging:\")\n",
    "pd.merge(random_df.loc[temp1_init_mask], random_df2.loc[temp2_init_mask], on = selected_cols_ran, how = \"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bdc3ea05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Actual df after iterative merging:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233438</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SW233426</td>\n",
       "      <td>229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233434             245        NaN        NaN        NaN        NaN   \n",
       "2  NW233438             231        4.0        1.0        1.0        1.0   \n",
       "3  SW233426             229        NaN        NaN        NaN        NaN   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  \\\n",
       "0                2.0        7.0        1.0       66.0       97.0   \n",
       "1                NaN       75.0        3.0        1.0        4.0   \n",
       "2                5.0        NaN        NaN        NaN        NaN   \n",
       "3                NaN       26.0      356.0       42.0       74.0   \n",
       "\n",
       "   Rep2_Deletions_BS  \n",
       "0               12.0  \n",
       "1                7.0  \n",
       "2                NaN  \n",
       "3                6.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## View output from iterative merging\n",
    "for df in df_list_ran[1:]:\n",
    "    if not df.empty:\n",
    "        colnames = df.columns.tolist()\n",
    "        mask = filtertsv.create_mask(df, colnames)\n",
    "        df = df.loc[mask]\n",
    "        df_full_ran = pd.merge(df_full_ran, df, on = selected_cols_ran, how = \"outer\")\n",
    "\n",
    "print(\"Actual df after iterative merging:\")\n",
    "df_full_ran"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7575d798",
   "metadata": {},
   "source": [
    "### Using test datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dc5810d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING TSV_FOLDER FOR SUBFOLDER: 7KO-Cyto\n",
      "Expected: Directory for individual_tsv folder\n",
      "Actual: c:\\Users\\Sonia Ling\\Desktop\\calculate_dr\\clean_tsv\\testing\\calculations\\7KO-Cyto\\individual_tsv\n",
      "\n",
      "NOW PRINTING TSV_LIST FOR SUBFOLDER: 7KO-Cyto\n",
      "Expected amount of files: 6\n",
      "Actual: 6\n",
      "\n",
      "NOW PRINTING TSV_FOLDER FOR SUBFOLDER: 7KO-Nuc\n",
      "Expected: Directory for individual_tsv folder\n",
      "Actual: c:\\Users\\Sonia Ling\\Desktop\\calculate_dr\\clean_tsv\\testing\\calculations\\7KO-Nuc\\individual_tsv\n",
      "\n",
      "NOW PRINTING TSV_LIST FOR SUBFOLDER: 7KO-Nuc\n",
      "Expected amount of files: 6\n",
      "Actual: 6\n"
     ]
    }
   ],
   "source": [
    "## Verifying tsv_folder and tsv_list\n",
    "for subfolder in input_dir.iterdir():\n",
    "    tsv_folder = input_dir/subfolder/\"individual_tsv\"\n",
    "\n",
    "    print(f\"\\nNOW PRINTING TSV_FOLDER FOR SUBFOLDER: {subfolder.name}\", \n",
    "          \"Expected: Directory for individual_tsv folder\", \n",
    "          f\"Actual: {tsv_folder}\", sep = \"\\n\")\n",
    "    \n",
    "    if subfolder.is_dir():\n",
    "        tsv_list = sorted(\n",
    "            tsv_folder.glob(\"*.tsv\"),\n",
    "            key = lambda x: int(re.search(r\"Rep(\\d+)\", x.name).group(1))\n",
    "        )\n",
    "\n",
    "        print(f\"\\nNOW PRINTING TSV_LIST FOR SUBFOLDER: {subfolder.name}\")\n",
    "        if (subfolder.name) == \"7KO-Cyto\":\n",
    "            print(\"Expected amount of files: 6\")\n",
    "        else:\n",
    "            print(\"Expected amount of files: 6\")\n",
    "        print(f\"Actual: {len(tsv_list)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "295d0f69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected: Columns should span from 'TranscriptID' to 'fit_b'\n",
      "Actual:\n",
      "['TranscriptID', 'Motif', 'Region', 'Chrom', 'Strand', 'TranscriptPosStart', 'TranscriptPosEnd', 'TranscriptModBase', 'GenomicModBase', 'TranscriptLength', 'DistFromAUG', 'DistFromSTOP', 'DistFromExonStart', 'DistFromExonEnd', 'fit_c', 'fit_s', 'fit_b']\n"
     ]
    }
   ],
   "source": [
    "## Test if selected column names are correct for 7KO-Cyto subfolder\n",
    "tsv_folder = input_dir/\"7KO-Cyto/individual_tsv\"\n",
    "\n",
    "tsv_list = sorted(\n",
    "    tsv_folder.glob(\"*.tsv\"),\n",
    "    key = lambda x: int(re.search(r\"Rep(\\d+)\", x.name).group(1))\n",
    ")\n",
    "\n",
    "df_list = [pd.read_csv(str(file), sep = \"\\t\") for file in tsv_list]\n",
    "df1_colnames = df_list[0].columns.tolist()\n",
    "selected_colnames = df1_colnames[0:17]\n",
    "\n",
    "print(\"Expected: Columns should span from 'TranscriptID' to 'fit_b'\",\n",
    "      \"Actual:\", selected_colnames, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1e16b46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TranscriptID',\n",
       " 'Motif',\n",
       " 'Region',\n",
       " 'Chrom',\n",
       " 'Strand',\n",
       " 'TranscriptPosStart',\n",
       " 'TranscriptPosEnd',\n",
       " 'TranscriptModBase',\n",
       " 'GenomicModBase',\n",
       " 'TranscriptLength',\n",
       " 'DistFromAUG',\n",
       " 'DistFromSTOP',\n",
       " 'DistFromExonStart',\n",
       " 'DistFromExonEnd',\n",
       " 'fit_c',\n",
       " 'fit_s',\n",
       " 'fit_b',\n",
       " 'Rep1_A_BS',\n",
       " 'Rep1_C_BS',\n",
       " 'Rep1_G_BS',\n",
       " 'Rep1_T_BS',\n",
       " 'Rep1_Deletions_BS',\n",
       " 'Rep1_DeletionRate_BS',\n",
       " 'Rep1_RealRate_BS',\n",
       " 'TotalCoverage']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e58b3bfb",
   "metadata": {},
   "outputs": [
    {
     "ename": "MergeError",
     "evalue": "Passing 'suffixes' which cause duplicate columns {'TotalCoverage_x'} is not allowed.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMergeError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[18], line 9\u001b[0m\n\u001b[0;32m      7\u001b[0m mask \u001b[38;5;241m=\u001b[39m filtertsv\u001b[38;5;241m.\u001b[39mcreate_mask(df, colnames)\n\u001b[0;32m      8\u001b[0m df \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mloc[mask]\n\u001b[1;32m----> 9\u001b[0m df_full \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(df_full, df, on \u001b[38;5;241m=\u001b[39m selected_colnames, how \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mouter\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\Sonia Ling\\anaconda3\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:184\u001b[0m, in \u001b[0;36mmerge\u001b[1;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[0;32m    169\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    170\u001b[0m     op \u001b[38;5;241m=\u001b[39m _MergeOperation(\n\u001b[0;32m    171\u001b[0m         left_df,\n\u001b[0;32m    172\u001b[0m         right_df,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    182\u001b[0m         validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[0;32m    183\u001b[0m     )\n\u001b[1;32m--> 184\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m op\u001b[38;5;241m.\u001b[39mget_result(copy\u001b[38;5;241m=\u001b[39mcopy)\n",
      "File \u001b[1;32mc:\\Users\\Sonia Ling\\anaconda3\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:888\u001b[0m, in \u001b[0;36m_MergeOperation.get_result\u001b[1;34m(self, copy)\u001b[0m\n\u001b[0;32m    884\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_indicator_pre_merge(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright)\n\u001b[0;32m    886\u001b[0m join_index, left_indexer, right_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_join_info()\n\u001b[1;32m--> 888\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_and_concat(\n\u001b[0;32m    889\u001b[0m     join_index, left_indexer, right_indexer, copy\u001b[38;5;241m=\u001b[39mcopy\n\u001b[0;32m    890\u001b[0m )\n\u001b[0;32m    891\u001b[0m result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_merge_type)\n\u001b[0;32m    893\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindicator:\n",
      "File \u001b[1;32mc:\\Users\\Sonia Ling\\anaconda3\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:840\u001b[0m, in \u001b[0;36m_MergeOperation._reindex_and_concat\u001b[1;34m(self, join_index, left_indexer, right_indexer, copy)\u001b[0m\n\u001b[0;32m    837\u001b[0m left \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft[:]\n\u001b[0;32m    838\u001b[0m right \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright[:]\n\u001b[1;32m--> 840\u001b[0m llabels, rlabels \u001b[38;5;241m=\u001b[39m _items_overlap_with_suffix(\n\u001b[0;32m    841\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft\u001b[38;5;241m.\u001b[39m_info_axis, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright\u001b[38;5;241m.\u001b[39m_info_axis, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msuffixes\n\u001b[0;32m    842\u001b[0m )\n\u001b[0;32m    844\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m left_indexer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_range_indexer(left_indexer, \u001b[38;5;28mlen\u001b[39m(left)):\n\u001b[0;32m    845\u001b[0m     \u001b[38;5;66;03m# Pinning the index here (and in the right code just below) is not\u001b[39;00m\n\u001b[0;32m    846\u001b[0m     \u001b[38;5;66;03m#  necessary, but makes the `.take` more performant if we have e.g.\u001b[39;00m\n\u001b[0;32m    847\u001b[0m     \u001b[38;5;66;03m#  a MultiIndex for left.index.\u001b[39;00m\n\u001b[0;32m    848\u001b[0m     lmgr \u001b[38;5;241m=\u001b[39m left\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mreindex_indexer(\n\u001b[0;32m    849\u001b[0m         join_index,\n\u001b[0;32m    850\u001b[0m         left_indexer,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    855\u001b[0m         use_na_proxy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[0;32m    856\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Users\\Sonia Ling\\anaconda3\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:2757\u001b[0m, in \u001b[0;36m_items_overlap_with_suffix\u001b[1;34m(left, right, suffixes)\u001b[0m\n\u001b[0;32m   2755\u001b[0m     dups\u001b[38;5;241m.\u001b[39mextend(rlabels[(rlabels\u001b[38;5;241m.\u001b[39mduplicated()) \u001b[38;5;241m&\u001b[39m (\u001b[38;5;241m~\u001b[39mright\u001b[38;5;241m.\u001b[39mduplicated())]\u001b[38;5;241m.\u001b[39mtolist())\n\u001b[0;32m   2756\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m dups:\n\u001b[1;32m-> 2757\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MergeError(\n\u001b[0;32m   2758\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPassing \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msuffixes\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m which cause duplicate columns \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mset\u001b[39m(dups)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   2759\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnot allowed.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   2760\u001b[0m     )\n\u001b[0;32m   2762\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m llabels, rlabels\n",
      "\u001b[1;31mMergeError\u001b[0m: Passing 'suffixes' which cause duplicate columns {'TotalCoverage_x'} is not allowed."
     ]
    }
   ],
   "source": [
    "init_mask = filtertsv.create_mask(df_list[0], df1_colnames)\n",
    "df_full = df_list[0].loc[init_mask]\n",
    "\n",
    "for df in df_list[1:]:\n",
    "    if not df.empty:\n",
    "        colnames = df.columns.tolist()\n",
    "        mask = filtertsv.create_mask(df, colnames)\n",
    "        df = df.loc[mask]\n",
    "        df_full = pd.merge(df_full, df, on = selected_colnames, how = \"outer\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dbf0aa0",
   "metadata": {},
   "source": [
    "## Testing Fisher's Exact Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e235c9d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_c</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep1_DeletionRate_BS</th>\n",
       "      <th>Rep1_RealRate_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_152686.4-2</td>\n",
       "      <td>UGUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_021159996.1</td>\n",
       "      <td>-</td>\n",
       "      <td>2687</td>\n",
       "      <td>2692</td>\n",
       "      <td>2689</td>\n",
       "      <td>3144</td>\n",
       "      <td>5102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981073</td>\n",
       "      <td>0.931667</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.758557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_003051.4-2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_009646196.1</td>\n",
       "      <td>-</td>\n",
       "      <td>2285</td>\n",
       "      <td>2290</td>\n",
       "      <td>2287</td>\n",
       "      <td>3891</td>\n",
       "      <td>3753</td>\n",
       "      <td>...</td>\n",
       "      <td>0.972996</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.140260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_003051.4-2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_009646196.1</td>\n",
       "      <td>-</td>\n",
       "      <td>2285</td>\n",
       "      <td>2290</td>\n",
       "      <td>2287</td>\n",
       "      <td>3891</td>\n",
       "      <td>3753</td>\n",
       "      <td>...</td>\n",
       "      <td>0.972996</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.140260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_003051.4-2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_009646196.1</td>\n",
       "      <td>-</td>\n",
       "      <td>2285</td>\n",
       "      <td>2290</td>\n",
       "      <td>2287</td>\n",
       "      <td>3891</td>\n",
       "      <td>3753</td>\n",
       "      <td>...</td>\n",
       "      <td>0.972996</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.140260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_001166496.2-2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NW_009646196.1</td>\n",
       "      <td>-</td>\n",
       "      <td>3044</td>\n",
       "      <td>3049</td>\n",
       "      <td>3046</td>\n",
       "      <td>3891</td>\n",
       "      <td>4512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.972996</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.140260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497402</th>\n",
       "      <td>rna-NM_004501.3</td>\n",
       "      <td>UUUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>4885</td>\n",
       "      <td>4890</td>\n",
       "      <td>4887</td>\n",
       "      <td>244852201</td>\n",
       "      <td>6789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.967359</td>\n",
       "      <td>0.866394</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.318686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497403</th>\n",
       "      <td>rna-NM_001142572.2</td>\n",
       "      <td>UGUAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>1234</td>\n",
       "      <td>1239</td>\n",
       "      <td>1236</td>\n",
       "      <td>247100431</td>\n",
       "      <td>1706</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981073</td>\n",
       "      <td>0.931667</td>\n",
       "      <td>0.001</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.578884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497404</th>\n",
       "      <td>rna-NM_001142572.2</td>\n",
       "      <td>UGUAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>1234</td>\n",
       "      <td>1239</td>\n",
       "      <td>1236</td>\n",
       "      <td>247100431</td>\n",
       "      <td>1706</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981073</td>\n",
       "      <td>0.931667</td>\n",
       "      <td>0.001</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.578884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497405</th>\n",
       "      <td>rna-NM_024804.3</td>\n",
       "      <td>UGUAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>1492</td>\n",
       "      <td>1497</td>\n",
       "      <td>1494</td>\n",
       "      <td>247100431</td>\n",
       "      <td>1964</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981073</td>\n",
       "      <td>0.931667</td>\n",
       "      <td>0.001</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.578884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497406</th>\n",
       "      <td>rna-NM_024804.3</td>\n",
       "      <td>UGUAG</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>1492</td>\n",
       "      <td>1497</td>\n",
       "      <td>1494</td>\n",
       "      <td>247100431</td>\n",
       "      <td>1964</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981073</td>\n",
       "      <td>0.931667</td>\n",
       "      <td>0.001</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.578884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>497407 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                TranscriptID  Motif Region           Chrom Strand  \\\n",
       "0          rna-NM_152686.4-2  UGUAG   3UTR  NW_021159996.1      -   \n",
       "1          rna-NM_003051.4-2  UUUAG   3UTR  NW_009646196.1      -   \n",
       "2          rna-NM_003051.4-2  UUUAG   3UTR  NW_009646196.1      -   \n",
       "3          rna-NM_003051.4-2  UUUAG   3UTR  NW_009646196.1      -   \n",
       "4       rna-NM_001166496.2-2  UUUAG   3UTR  NW_009646196.1      -   \n",
       "...                      ...    ...    ...             ...    ...   \n",
       "497402       rna-NM_004501.3  UUUAA   3UTR    NC_000001.11      -   \n",
       "497403    rna-NM_001142572.2  UGUAG    CDS    NC_000001.11      -   \n",
       "497404    rna-NM_001142572.2  UGUAG    CDS    NC_000001.11      -   \n",
       "497405       rna-NM_024804.3  UGUAG    CDS    NC_000001.11      -   \n",
       "497406       rna-NM_024804.3  UGUAG    CDS    NC_000001.11      -   \n",
       "\n",
       "        TranscriptPosStart  TranscriptPosEnd  TranscriptModBase  \\\n",
       "0                     2687              2692               2689   \n",
       "1                     2285              2290               2287   \n",
       "2                     2285              2290               2287   \n",
       "3                     2285              2290               2287   \n",
       "4                     3044              3049               3046   \n",
       "...                    ...               ...                ...   \n",
       "497402                4885              4890               4887   \n",
       "497403                1234              1239               1236   \n",
       "497404                1234              1239               1236   \n",
       "497405                1492              1497               1494   \n",
       "497406                1492              1497               1494   \n",
       "\n",
       "        GenomicModBase  TranscriptLength  ...     fit_c     fit_s  fit_b  \\\n",
       "0                 3144              5102  ...  0.981073  0.931667  0.001   \n",
       "1                 3891              3753  ...  0.972996  0.901845  0.001   \n",
       "2                 3891              3753  ...  0.972996  0.901845  0.001   \n",
       "3                 3891              3753  ...  0.972996  0.901845  0.001   \n",
       "4                 3891              4512  ...  0.972996  0.901845  0.001   \n",
       "...                ...               ...  ...       ...       ...    ...   \n",
       "497402       244852201              6789  ...  0.967359  0.866394  0.001   \n",
       "497403       247100431              1706  ...  0.981073  0.931667  0.001   \n",
       "497404       247100431              1706  ...  0.981073  0.931667  0.001   \n",
       "497405       247100431              1964  ...  0.981073  0.931667  0.001   \n",
       "497406       247100431              1964  ...  0.981073  0.931667  0.001   \n",
       "\n",
       "        Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  Rep1_Deletions_BS  \\\n",
       "0             5.0        0.0        0.0        0.0                1.0   \n",
       "1           121.0        0.0        0.0        0.0                2.0   \n",
       "2           121.0        0.0        0.0        0.0                2.0   \n",
       "3           121.0        0.0        0.0        0.0                2.0   \n",
       "4           121.0        0.0        0.0        0.0                2.0   \n",
       "...           ...        ...        ...        ...                ...   \n",
       "497402       33.0        0.0        0.0        0.0                2.0   \n",
       "497403       11.0        0.0        0.0        0.0                1.0   \n",
       "497404       11.0        0.0        0.0        0.0                1.0   \n",
       "497405       11.0        0.0        0.0        0.0                1.0   \n",
       "497406       11.0        0.0        0.0        0.0                1.0   \n",
       "\n",
       "        Rep1_DeletionRate_BS  Rep1_RealRate_BS  \n",
       "0                   0.166667          0.758557  \n",
       "1                   0.016260          0.140260  \n",
       "2                   0.016260          0.140260  \n",
       "3                   0.016260          0.140260  \n",
       "4                   0.016260          0.140260  \n",
       "...                      ...               ...  \n",
       "497402              0.057143          0.318686  \n",
       "497403              0.083333          0.578884  \n",
       "497404              0.083333          0.578884  \n",
       "497405              0.083333          0.578884  \n",
       "497406              0.083333          0.578884  \n",
       "\n",
       "[497407 rows x 24 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "11a933a4",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not DataFrame",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[21], line 11\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;124;03mPART II: Iteratively merge remaining dfs\u001b[39;00m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;124;03m* The range [1:] means 'Start from 2nd item in list, and continue \u001b[39;00m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;124;03m    looping until you reach the end'\u001b[39;00m\n\u001b[0;32m      8\u001b[0m \u001b[38;5;124;03m* Output is df_full, which contains rows from all merged dfs\u001b[39;00m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     10\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m df_list[\u001b[38;5;241m1\u001b[39m:]:\n\u001b[1;32m---> 11\u001b[0m     colnames \u001b[38;5;241m=\u001b[39m df_list[i]\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[0;32m     12\u001b[0m     mask \u001b[38;5;241m=\u001b[39m filtertsv\u001b[38;5;241m.\u001b[39mcreate_mask(df_list[i], colnames)\n\u001b[0;32m     13\u001b[0m     df_list[i] \u001b[38;5;241m=\u001b[39m df_list[i]\u001b[38;5;241m.\u001b[39mloc[mask]\n",
      "\u001b[1;31mTypeError\u001b[0m: list indices must be integers or slices, not DataFrame"
     ]
    }
   ],
   "source": [
    "# ## Sort column names\n",
    "# \"\"\"\n",
    "# PART I: Collect all column names\n",
    "# \"\"\"\n",
    "# merged_colnames = df_full.columns.tolist()\n",
    "\n",
    "# \"\"\"\n",
    "# PART II: Sort by replicate order\n",
    "# * Select columns that contain \"Rep\\d+\", such as Rep1, Rep2, Rep3, etc.,\n",
    "#     and put them in a list\n",
    "# * Use set() to remove duplicates, since sets can only contain unique vals\n",
    "# * On those columns, select the first RegEx match group. In this case,\n",
    "#     it'd be \"\\d+\" or the digit, such as 1, 2, 3, etc.\n",
    "# * Sort columns by those digits in ascending order using sorted() \n",
    "# \"\"\"\n",
    "# rep_list = sorted(\n",
    "#     set([re.search(r\"(Rep\\d+)\", col).group(1) for col in merged_colnames \n",
    "#         if re.search(r\"(Rep\\d+)\", col)]), \n",
    "#         key = lambda x: int(re.search(r\"Rep(\\d+)\", x).group(1))\n",
    "# )\n",
    "\n",
    "# \"\"\"\n",
    "# NOTES:\n",
    "# * Drop nulls to ensure deletion rates for all 3 replicates\n",
    "# * Calculate p-vals with Fisher's Exact Test\n",
    "# * Sort DeletionRate by descending order\n",
    "# * Keep only first 50 rows\n",
    "# \"\"\"\n",
    "# ## Group corresponding BS/NBS into separate lists\n",
    "# df_merged = df_full.dropna()\n",
    "# for rep in rep_list:\n",
    "#     bs_base_pattern = re.compile(fr\"{rep}_(A|C|G|T)_BS$\")\n",
    "#     nbs_base_pattern = re.compile(fr\"{rep}_(A|C|G|T)_NBS$\")\n",
    "#     pattern_dict = {f\"{rep}_Bases_BS\": [col for col in merged_colnames \n",
    "#                                         if bs_base_pattern.match(col)],\n",
    "#                     f\"{rep}_Bases_NBS\": [col for col in merged_colnames \n",
    "#                                         if nbs_base_pattern.match(col)]}\n",
    "\n",
    "# ## Run Fisher's Exact Test (p-values)\n",
    "# filtertsv.merged_output(df_merged, rep_list, pattern_dict)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
