{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ebdf09f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import traceback\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from scipy.stats import fisher_exact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d48ade8",
   "metadata": {},
   "source": [
    "## Class initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f88fa575",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FilterTSV:\n",
    "   def create_mask(self, df, colnames):\n",
    "      \"\"\"\n",
    "      NOTES:\n",
    "      * Select columns that contain \"Deletions\" and put them in a list\n",
    "      * Use set() to remove duplicates, since sets can only contain unique vals\n",
    "      * Pass column names in list to dataframe to create a mask that drops rows\n",
    "        where Deletions == 0 and there are nulls\n",
    "      \"\"\"\n",
    "      del_list = set([col for col in colnames if re.search(r\"Deletions\", col)])\n",
    "      mask = (df[del_list] != 0).all(axis=1) & (~df.isnull().any(axis=1))\n",
    "      return mask\n",
    "\n",
    "   def merged_output(self, df_merged, rep_list, pattern_dict):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      1. Takes all columns from merged df and organizes them by BS/NBS type \n",
    "      2. Sums up corresponding bases/deletions & creates 4 new columns per replicate\n",
    "      3. Selects the new columns\n",
    "         * Reshapes each row into 2x2 matrix\n",
    "         * Runs Fisher's Exact Test\n",
    "         * Appends p-val column\n",
    "      \"\"\"\n",
    "      try:\n",
    "         for rep in rep_list:\n",
    "            ## Define names of summed BS/NBS columns\n",
    "            new_cols = [f\"{rep}_TotalBases_BS\", \n",
    "                        f\"{rep}_TotalBases_NBS\"]\n",
    "            \n",
    "            ## Define generic entries for 2x2 contingency table\n",
    "            fisher_cols = [f\"{rep}_TotalBases_BS\", \n",
    "                           f\"{rep}_Deletions_BS\", \n",
    "                           f\"{rep}_TotalBases_NBS\", \n",
    "                           f\"{rep}_Deletions_NBS\"]\n",
    "\n",
    "            ## Calculate p-values\n",
    "            \"\"\"\n",
    "            PART I: For each replicate, find total bases for BS/NBS\n",
    "            * Zips pattern_dict and new_cols together\n",
    "              -> Reminder: pattern_dict = {[List of BS colnames], [List of NBS colnames]}\n",
    "                           new_cols = [\"TotalBases_BS\", \"TotalBases_NBS\"]\n",
    "            * Sum of all entries in 1st list of pattern_dict \n",
    "              -> Stored under 1st colname in new_cols\n",
    "            * Sum of all entries in 2nd list of pattern_dict \n",
    "              -> Stored under 2nd colname in new_cols \n",
    "            \"\"\"\n",
    "            for col, key in zip(new_cols, pattern_dict):\n",
    "               if col not in df_merged.columns:\n",
    "                  df_merged[col] = df_merged[pattern_dict[key]].sum(axis=1)\n",
    "\n",
    "            \"\"\"\n",
    "            PART II: Run Fisher's Exact Test using 2x2 table\n",
    "            * For each row in df_merged:\n",
    "              -> Select the specified columns from fisher_cols\n",
    "              -> Reshape the 4 columns into separate 3D arrays of size 2x2\n",
    "                 (these will be our 2x2 tables)\n",
    "              -> Use these arrays for numpy batch processing\n",
    "            * Run Fisher's Exact Test (scipy) on each table, then select second result\n",
    "              of test AKA the p-val using 'fisher_exact(table)[1]'\n",
    "            \"\"\"\n",
    "            if set(fisher_cols).issubset(df_merged.columns):\n",
    "               df_merged = df_merged.dropna(subset = fisher_cols)\n",
    "               arr = df_merged[fisher_cols].values.reshape(-1, 2, 2) \n",
    "               pvals = [fisher_exact(table)[1] for table in arr]\n",
    "               df_merged[f\"{rep}_Pvalue\"] = pvals\n",
    "         return df_merged\n",
    "      except Exception as e:\n",
    "         print(f\"Failed to calculate p-value for {rep}: {e}\")\n",
    "         traceback.print_exc()\n",
    "         raise\n",
    "\n",
    "   def filter_means(self, df_filtered, colname, cols):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      * Use to filter by average (Cutoffs #4-6)\n",
    "      \"\"\"\n",
    "      ## Calculate average and standard deviation\n",
    "      df_filtered[colname] = df_filtered[cols].mean(axis = 1)      \n",
    "      std_colname = colname.replace(\"Avg\", \"Std\")\n",
    "      df_filtered[std_colname] = df_filtered[cols].std(axis = 1)\n",
    "\n",
    "      ## Sort by descending DeletionRate\n",
    "      df_filtered = df_filtered.sort_values(by = colname, \n",
    "                                            ascending = False)\n",
    "\n",
    "      ## If BS, apply filters to average columns\n",
    "      if \"_BS\" in colname:\n",
    "         if \"DeletionCt\" in colname:\n",
    "            df_filtered[colname] = df_filtered[colname].ge(5)\n",
    "         elif \"DeletionRate\" in colname:\n",
    "            df_filtered[colname] = df_filtered[colname].ge(0.02)\n",
    "\n",
    "      return df_filtered\n",
    "\n",
    "   def filtered_output(self, df_merged, rep_list):\n",
    "      \"\"\"\n",
    "      PURPOSE:\n",
    "      a) Adds cutoffs from BID-Pipe protocol:\n",
    "         1. Pvalue across all replicates <= 0.0001\n",
    "         2. RealRate across all replicates >= 0.3\n",
    "         3. Total sequencing coverage for each BS and NBS replicate >= 20\n",
    "         4. Average Deletions across all BS replicates >= 5\n",
    "         5. Average DeletionRate across all BS replicates >= 0.02\n",
    "         6. Average DeletionRate is 2x higher in BS compared to NBS\n",
    "      b) Saves filtered and discarded rows in separate dataframes\n",
    "      \"\"\"\n",
    "      try:\n",
    "         ## Cutoff 1: Pvalue\n",
    "         pval_list = [col for col in df_merged.columns \n",
    "                      if re.search(r\"Pvalue$\", col)]\n",
    "         cutoff1 = df_merged[pval_list].le(0.0001).all(axis=1)\n",
    "         df_filtered = df_merged.loc[cutoff1]\n",
    "\n",
    "         ## Cutoff 2: RealRate\n",
    "         realrate_list = [col for col in df_filtered.columns \n",
    "                          if re.search(r\"RealRate\", col)]\n",
    "         cutoff2 = df_filtered[realrate_list].ge(0.3).all(axis=1)\n",
    "         df_filtered = df_filtered.loc[cutoff2]\n",
    "\n",
    "         ## Cutoff 3: Total sequencing coverage\n",
    "         for rep in rep_list:\n",
    "            for sample in [\"BS\", \"NBS\"]:\n",
    "               coverage_list = [col for col in df_filtered.columns if \n",
    "                                re.match(fr\"{rep}_(TotalBases|Deletions)_{sample}\", col)]\n",
    "               total_sum = df_filtered[coverage_list].sum(axis = 1)\n",
    "               cutoff3 = total_sum.ge(20)\n",
    "               df_filtered = df_filtered.loc[cutoff3]\n",
    "\n",
    "         ## Cutoff 4: Average Deletions (BS)\n",
    "         avg_del_bs = \"AvgDeletionCt_BS\"\n",
    "         del_col_bs = [col for col in df_filtered.columns \n",
    "                       if re.search(r\"_Deletions_BS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_del_bs, del_col_bs)\n",
    "\n",
    "         ## Cutoff 5: Average DeletionRate (BS)\n",
    "         avg_dr_bs = \"AvgDeletionRate_BS\"\n",
    "         dr_col_bs = [col for col in df_filtered.columns \n",
    "                      if re.search(r\"_DeletionRate_BS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_dr_bs, dr_col_bs)\n",
    "\n",
    "         ## Cutoff 6: Average DeletionRate is 2x higher in BS compared to NBS\n",
    "         avg_dr_nbs = \"AvgDeletionRate_NBS\"\n",
    "         dr_col_nbs = [col for col in df_filtered.columns \n",
    "                       if re.search(r\"_DeletionRate_NBS$*\", col)]\n",
    "         df_filtered = self.filter_means(df_filtered, avg_dr_nbs, dr_col_nbs)\n",
    "         \n",
    "         cutoff6 = df_filtered[avg_dr_bs] >= 2 * df_filtered[avg_dr_nbs]\n",
    "         df_filtered = df_filtered[cutoff6]\n",
    "\n",
    "         print(\"Successfully applied cutoffs.\")\n",
    "\n",
    "         return df_filtered\n",
    "      \n",
    "      except Exception as e:\n",
    "         print(f\"Failed to apply cutoffs from BID-Pipe protocol: {e}\")\n",
    "         traceback.print_exc()\n",
    "         raise\n",
    "\n",
    "filtertsv = FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7213f1ce",
   "metadata": {},
   "source": [
    "## Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1a82247a",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_path = Path.cwd()\n",
    "input_dir = current_path/\"calculations\"\n",
    "filtertsv = FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba0541a",
   "metadata": {},
   "source": [
    "## Testing helper functions in FilterTSV()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d2dc3db",
   "metadata": {},
   "source": [
    "### Testing create_mask()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a2ad7424",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "1  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_Deletions_BS  \n",
       "0                  2                  5  \n",
       "1                  2                  5  \n",
       "2                  3                  7  \n",
       "3                  0                  5  \n",
       "4                  0                  0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Testing del_list variable\n",
    "\n",
    "random_list = [{\"Chrom\": \"NW233434\", \n",
    "                \"GenomicModBase\": 231, \n",
    "                \"Rep1_A_BS\": 4, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 2,\n",
    "                \"Rep2_Deletions_BS\": 5}, \n",
    "\n",
    "                {\"Chrom\": \"NW233434\", \n",
    "                \"GenomicModBase\": 231, \n",
    "                \"Rep1_A_BS\": 4, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 2,\n",
    "                \"Rep2_Deletions_BS\": 5},\n",
    "\n",
    "                {\"Chrom\": \"NW233435\", \n",
    "                 \"GenomicModBase\": 331, \n",
    "                 \"Rep1_A_BS\": None, \n",
    "                 \"Rep1_C_BS\": None, \n",
    "                 \"Rep1_G_BS\": None, \n",
    "                 \"Rep1_T_BS\": None, \n",
    "                 \"Rep1_Deletions_BS\": 3,\n",
    "                 \"Rep2_Deletions_BS\": 7},\n",
    "\n",
    "                {\"Chrom\": \"NW233433\", \n",
    "                \"GenomicModBase\": 230, \n",
    "                \"Rep1_A_BS\": 3, \n",
    "                \"Rep1_C_BS\": 1, \n",
    "                \"Rep1_G_BS\": 1, \n",
    "                \"Rep1_T_BS\": 1, \n",
    "                \"Rep1_Deletions_BS\": 0,\n",
    "                \"Rep2_Deletions_BS\": 5},\n",
    "                \n",
    "                {\"Chrom\": \"NW233432\", \n",
    "                \"GenomicModBase\": 229, \n",
    "                \"Rep1_A_BS\": 2, \n",
    "                \"Rep1_C_BS\": 6, \n",
    "                \"Rep1_G_BS\": 4, \n",
    "                \"Rep1_T_BS\": 7, \n",
    "                \"Rep1_Deletions_BS\": 0,\n",
    "                \"Rep2_Deletions_BS\": 0}]\n",
    "random_df = pd.DataFrame(random_list)\n",
    "\n",
    "random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3141b58d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_Deletions_BS  \n",
       "0                  2                  5  \n",
       "2                  3                  7  \n",
       "3                  0                  5  \n",
       "4                  0                  0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random_df = random_df.drop_duplicates()\n",
    "random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3b927b1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING DEL_LIST VARIABLE (order in list doesn't matter)\n",
      "Expected: ['Rep2_Deletions_BS', 'Rep1_Deletions_BS']\n",
      "Actual: ['Rep2_Deletions_BS', 'Rep1_Deletions_BS']\n"
     ]
    }
   ],
   "source": [
    "#### CHANGE: added list()\n",
    "del_list = list(set([col for col in random_df.columns if re.search(\"Deletions\", col)]))\n",
    "\n",
    "print(\"NOW TESTING DEL_LIST VARIABLE (order in list doesn't matter)\",\n",
    "      \"Expected: ['Rep2_Deletions_BS', 'Rep1_Deletions_BS']\",\n",
    "      f\"Actual: {del_list}\", sep = \"\\n\")\n",
    "\n",
    "## Testing mask variable\n",
    "#### CHANGE: fixed mask\n",
    "mask = ~(random_df[del_list] == 0).any(axis = 1) & (random_df.notna().all(axis = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b75b943e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING ORIGINAL RANDOM_DF\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NW233435</td>\n",
       "      <td>331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NW233433</td>\n",
       "      <td>230</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NW233432</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "2  NW233435             331        NaN        NaN        NaN        NaN   \n",
       "3  NW233433             230        3.0        1.0        1.0        1.0   \n",
       "4  NW233432             229        2.0        6.0        4.0        7.0   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_Deletions_BS  \n",
       "0                  2                  5  \n",
       "2                  3                  7  \n",
       "3                  0                  5  \n",
       "4                  0                  0  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"\\nNOW PRINTING ORIGINAL RANDOM_DF\")\n",
    "random_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6f4e8496",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING FILTERED RANDOM_DF\n",
      "Expected: Rows with 0 deletions or nulls should be removed\n",
      "Actual:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>Rep1_A_BS</th>\n",
       "      <th>Rep1_C_BS</th>\n",
       "      <th>Rep1_G_BS</th>\n",
       "      <th>Rep1_T_BS</th>\n",
       "      <th>Rep1_Deletions_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NW233434</td>\n",
       "      <td>231</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom  GenomicModBase  Rep1_A_BS  Rep1_C_BS  Rep1_G_BS  Rep1_T_BS  \\\n",
       "0  NW233434             231        4.0        1.0        1.0        1.0   \n",
       "\n",
       "   Rep1_Deletions_BS  Rep2_Deletions_BS  \n",
       "0                  2                  5  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"\\nNOW PRINTING FILTERED RANDOM_DF\",\n",
    "      \"Expected: Rows with 0 deletions or nulls should be removed\",\n",
    "      \"Actual:\", sep = \"\\n\")\n",
    "good_df = random_df.loc[mask]\n",
    "good_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199b08e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXPECTED: Failed to filter out DeletionRate == 0 in bad_df\n",
      "EXPECTED: Successfully filtered out DeletionRate == 0 in good_df\n"
     ]
    }
   ],
   "source": [
    "bad_mask = (random_df[random_df[del_list] != 0].all(axis = 1)) & (~random_df.isnull().any(axis = 1))\n",
    "bad_df = random_df.loc[bad_mask]\n",
    "\n",
    "if 0 in bad_df[del_list].values:\n",
    "    print(\"EXPECTED: Failed to filter out DeletionRate == 0 in bad_df\")\n",
    "else:\n",
    "    print(\"NOT EXPECTED: Successfully filtered out DeletionRate == 0 in bad_df\")\n",
    "\n",
    "if 0 in good_df[del_list].values:\n",
    "    print(\"NOT EXPECTED: Failed to filter out DeletionRate == 0 in good_df\")\n",
    "else:\n",
    "    print(\"EXPECTED: Successfully filtered out DeletionRate == 0 in good_df\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a02ce8df",
   "metadata": {},
   "source": [
    "## Testing iterative merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dc5810d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOW PRINTING TSV_FOLDER FOR SUBFOLDER: 7KO-Cyto\n",
      "Expected: Directory for individual_tsv folder\n",
      "Actual: c:\\Users\\Sonia Ling\\Desktop\\calculate_dr\\clean_tsv\\testing\\calculations\\7KO-Cyto\\individual_tsv\n",
      "\n",
      "NOW PRINTING TSV_LIST FOR SUBFOLDER: 7KO-Cyto\n",
      "Expected amount of files: 2\n",
      "Actual: 2\n",
      "\n",
      "NOW PRINTING TSV_FOLDER FOR SUBFOLDER: WT-Cyto\n",
      "Expected: Directory for individual_tsv folder\n",
      "Actual: c:\\Users\\Sonia Ling\\Desktop\\calculate_dr\\clean_tsv\\testing\\calculations\\WT-Cyto\\individual_tsv\n",
      "\n",
      "NOW PRINTING TSV_LIST FOR SUBFOLDER: WT-Cyto\n",
      "Expected amount of files: 1\n",
      "Actual: 1\n"
     ]
    }
   ],
   "source": [
    "## Verifying tsv_folder and tsv_list\n",
    "for subfolder in input_dir.iterdir():\n",
    "    tsv_folder = input_dir/subfolder/\"individual_tsv\"\n",
    "\n",
    "    print(f\"\\nNOW PRINTING TSV_FOLDER FOR SUBFOLDER: {subfolder.name}\", \n",
    "          \"Expected: Directory for individual_tsv folder\", \n",
    "          f\"Actual: {tsv_folder}\", sep = \"\\n\")\n",
    "    \n",
    "    if subfolder.is_dir():\n",
    "        tsv_list = sorted(\n",
    "            tsv_folder.glob(\"*.tsv\"),\n",
    "            key = lambda x: int(re.search(r\"Rep(\\d+)\", x.name).group(1))\n",
    "        )\n",
    "\n",
    "        print(f\"\\nNOW PRINTING TSV_LIST FOR SUBFOLDER: {subfolder.name}\")\n",
    "        if (subfolder.name) == \"7KO-Cyto\":\n",
    "            print(\"Expected amount of files: 2\")\n",
    "        else:\n",
    "            print(\"Expected amount of files: 1\")\n",
    "        print(f\"Actual: {len(tsv_list)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "295d0f69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TranscriptID', 'Motif', 'Region', 'Chrom', 'Strand', 'TranscriptPosStart', 'TranscriptPosEnd', 'TranscriptModBase', 'GenomicModBase', 'TranscriptLength', 'DistFromAUG', 'DistFromSTOP', 'DistFromExonStart', 'DistFromExonEnd', 'fit_c', 'fit_s', 'fit_b', 'Rep1_A_BS', 'Rep1_C_BS', 'Rep1_G_BS', 'Rep1_T_BS', 'Rep1_Deletions_BS', 'Rep1_DeletionRate_BS', 'Rep1_RealRate_BS']\n"
     ]
    }
   ],
   "source": [
    "## For simplicity's sake, we'll stick to the 7KO-Cyto subfolder for testing\n",
    "tsv_folder = input_dir/\"7KO-Cyto/individual_tsv\"\n",
    "\n",
    "tsv_list = sorted(\n",
    "    tsv_folder.glob(\"*.tsv\"),\n",
    "    key = lambda x: int(re.search(r\"Rep(\\d+)\", x.name).group(1))\n",
    ")\n",
    "\n",
    "df_list = [pd.read_csv(str(file), sep = \"\\t\") for file in tsv_list]\n",
    "df1_colnames = df_list[0].columns.tolist()\n",
    "\n",
    "print(df1_colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1eb8853d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# selected_colnames = df1_colnames[0:17]\n",
    "# init_mask = filtertsv.create_mask(df_list[0], df1_colnames)\n",
    "# # df_full = df_list[0].loc[init_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "11a933a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    # df1_colnames = df_list[0].columns.tolist()\n",
    "    # selected_colnames = df1_colnames[0:17]\n",
    "    # init_mask = filtertsv.create_mask(df_list[0], df1_colnames)\n",
    "    # df_full = df_list[0].loc[init_mask]\n",
    "\n",
    "    # \"\"\"\n",
    "    # PART II: Iteratively merge remaining dfs\n",
    "    # * The range [1:] means 'Start from 2nd item in list, and continue \n",
    "    #     looping until you reach the end'\n",
    "    # * Output is df_full, which contains rows from all merged dfs\n",
    "    # \"\"\"\n",
    "    # for i in df_list[1:]:\n",
    "    #     colnames = df_list[i].columns.tolist()\n",
    "    #     mask = filtertsv.create_mask(df_list[i], colnames)\n",
    "    #     df_list[i] = df_list[i].loc[mask]\n",
    "    #     df_full = pd.merge(df_full, df_list[i], on = selected_colnames, how = \"outer\")\n",
    "\n",
    "    # ## Sort column names\n",
    "    # \"\"\"\n",
    "    # PART I: Collect all column names\n",
    "    # \"\"\"\n",
    "    # merged_colnames = df_full.columns.tolist()\n",
    "\n",
    "    # \"\"\"\n",
    "    # PART II: Sort by replicate order\n",
    "    # * Select columns that contain \"Rep\\d+\", such as Rep1, Rep2, Rep3, etc.,\n",
    "    #     and put them in a list\n",
    "    # * Use set() to remove duplicates, since sets can only contain unique vals\n",
    "    # * On those columns, select the first RegEx match group. In this case,\n",
    "    #     it'd be \"\\d+\" or the digit, such as 1, 2, 3, etc.\n",
    "    # * Sort columns by those digits in ascending order using sorted() \n",
    "    # \"\"\"\n",
    "    # rep_list = sorted(\n",
    "    #     set([re.search(r\"(Rep\\d+)\", col).group(1) for col in merged_colnames \n",
    "    #         if re.search(r\"(Rep\\d+)\", col)]), \n",
    "    #         key = lambda x: int(re.search(r\"Rep(\\d+)\", x).group(1))\n",
    "    # )\n",
    "\n",
    "    # \"\"\"\n",
    "    # NOTES:\n",
    "    # * Drop nulls to ensure deletion rates for all 3 replicates\n",
    "    # * Calculate p-vals with Fisher's Exact Test\n",
    "    # * Sort DeletionRate by descending order\n",
    "    # * Keep only first 50 rows\n",
    "    # \"\"\"\n",
    "    # ## Group corresponding BS/NBS into separate lists\n",
    "    # df_merged = df_full.dropna()\n",
    "    # for rep in rep_list:\n",
    "    #     bs_base_pattern = re.compile(fr\"{rep}_(A|C|G|T)_BS$\")\n",
    "    #     nbs_base_pattern = re.compile(fr\"{rep}_(A|C|G|T)_NBS$\")\n",
    "    #     pattern_dict = {f\"{rep}_Bases_BS\": [col for col in merged_colnames \n",
    "    #                                         if bs_base_pattern.match(col)],\n",
    "    #                     f\"{rep}_Bases_NBS\": [col for col in merged_colnames \n",
    "    #                                         if nbs_base_pattern.match(col)]}\n",
    "\n",
    "    # ## Run Fisher's Exact Test (p-values)\n",
    "    # filtertsv.merged_output(df_merged, rep_list, pattern_dict)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
