{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "626ab5ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import traceback\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "## Disable .loc indexing warning\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbf26e5a",
   "metadata": {},
   "source": [
    "### Loading in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e45fa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_s</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "      <th>Rep2_DeletionRate_BS</th>\n",
       "      <th>Rep2_RealRate_BS</th>\n",
       "      <th>TotalCoverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-XM_005245453.2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3857</td>\n",
       "      <td>3862</td>\n",
       "      <td>3859</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_022731.5</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3860</td>\n",
       "      <td>3865</td>\n",
       "      <td>3862</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.901845</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_012458.4</td>\n",
       "      <td>UAUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>-</td>\n",
       "      <td>1250</td>\n",
       "      <td>1255</td>\n",
       "      <td>1252</td>\n",
       "      <td>2426036</td>\n",
       "      <td>1664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.966038</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.735916</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001322267.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1730</td>\n",
       "      <td>1735</td>\n",
       "      <td>1732</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_004697.5</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1691</td>\n",
       "      <td>1696</td>\n",
       "      <td>1693</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890419</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TranscriptID  Motif Region         Chrom Strand  TranscriptPosStart  \\\n",
       "0  rna-XM_005245453.2  UUUAG   3UTR  NC_000001.11      -                3857   \n",
       "1     rna-NM_022731.5  UUUAG   3UTR  NC_000001.11      -                3860   \n",
       "2     rna-NM_012458.4  UAUAG   3UTR  NC_000019.10      -                1250   \n",
       "3  rna-NM_001322267.2  UCUAA   3UTR  NC_000009.12      +                1730   \n",
       "4     rna-NM_004697.5  UCUAA   3UTR  NC_000009.12      +                1691   \n",
       "\n",
       "   TranscriptPosEnd  TranscriptModBase  GenomicModBase  TranscriptLength  ...  \\\n",
       "0              3862               3859       205715358              6396  ...   \n",
       "1              3865               3862       205715358              6399  ...   \n",
       "2              1255               1252         2426036              1664  ...   \n",
       "3              1735               1732       113291699              2939  ...   \n",
       "4              1696               1693       113291699              2900  ...   \n",
       "\n",
       "      fit_s  fit_b  Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  \\\n",
       "0  0.901845  0.001        249          0          0          0   \n",
       "1  0.901845  0.001        249          0          0          0   \n",
       "2  0.966038  0.001          8         15          0          0   \n",
       "3  0.890419  0.001          0          0          0         25   \n",
       "4  0.890419  0.001          0          0          0         25   \n",
       "\n",
       "   Rep2_Deletions_BS  Rep2_DeletionRate_BS  Rep2_RealRate_BS  TotalCoverage  \n",
       "0                 24              0.087912          0.506265            273  \n",
       "1                 24              0.087912          0.506265            273  \n",
       "2                  2              0.080000          0.735916             25  \n",
       "3                  2              0.074074          0.426971             27  \n",
       "4                  2              0.074074          0.426971             27  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load in file\n",
    "current_path = Path.cwd()\n",
    "df = pd.read_csv(current_path/\"KEH-Rep2-7KO-HEK293T-Cyto-BS.sorted.tsv\", sep = \"\\t\")\n",
    "\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8c6689f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Columns that are always the same across all TSVs, excluding the best-fit values\n",
    "constant_cols = df.columns.tolist()[0:14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e7635866",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clean = df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a6d28806",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\n",
      "7KO-Cyto-BS_processed_fastqs\n"
     ]
    }
   ],
   "source": [
    "subfolder = \"KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\"\n",
    "folder_name = \"7KO-Cyto-BS_processed_fastqs\"\n",
    "print(subfolder, folder_name, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dd23f9c",
   "metadata": {},
   "source": [
    "### Testing make_key()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "820cabb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING REP_LIST\n",
      "Expected: ['Rep1', 'Rep2', 'Rep3'], Actual: ['Rep1', 'Rep2', 'Rep3']\n",
      "\n",
      "NOW TESTING REP_LIST\n",
      "Expected: ['Rep1', 'Rep5', 'Rep10'], Actual: ['Rep1', 'Rep5', 'Rep10']\n"
     ]
    }
   ],
   "source": [
    "## Function definition\n",
    "def make_key(subfolder, base_key):\n",
    "    \"\"\"\n",
    "    PURPOSE:\n",
    "    Modifies names of dictionary keys based on the Rep # (detected via RegEx)\n",
    "    and Sample Type (BS, NBS) in a given subfolder name.\n",
    "    ---\n",
    "    NOTES:\n",
    "    * sorted(set(rep_matches)): Removes duplicate reps, sorts in ascending order\n",
    "    * for rep in rep_list: Adds replicate prefix to dict key names\n",
    "    * for sample in ['BS', 'NBS']: Adds sample type suffix to dict key names\n",
    "    \"\"\"\n",
    "    rep_matches = re.findall(r\"Rep\\d+\", str(subfolder))\n",
    "    rep_list = sorted(set(rep_matches))\n",
    "    \n",
    "    for rep in rep_list:\n",
    "        if f\"-{rep}-\" in str(subfolder):\n",
    "            prefix = rep + \"_\"\n",
    "            break\n",
    "    \n",
    "    for sample in [\"BS\", \"NBS\"]:\n",
    "        if f\"-{sample}_\" in str(subfolder):\n",
    "            suffix = \"_\" + sample\n",
    "            break\n",
    "    \n",
    "    return prefix + base_key + suffix\n",
    "\n",
    "\"\"\"\n",
    "PURPOSE:\n",
    "* Showcases how rep_list is supposed to behave\n",
    "* Reps should be listed in integer order\n",
    "---\n",
    "NOTES: \n",
    "* These variables are not used outside of\n",
    "  this cell; for demonstration purposes only\n",
    "\"\"\"\n",
    "\n",
    "## Test 1: Rep1, Rep2, Rep3\n",
    "folder_list = [\"KEH-Rep3-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\"]\n",
    "rep_matches = re.findall(r\"Rep\\d+\", str(folder_list))\n",
    "rep_list = sorted(set(rep_matches), key = lambda x: int(x[3:]))\n",
    "\n",
    "print(\"NOW TESTING REP_LIST\")\n",
    "print(f\"Expected: ['Rep1', 'Rep2', 'Rep3'], Actual: {rep_list}\")\n",
    "\n",
    "## Test 2: Rep1, Rep5, Rep10\n",
    "folder_list = [\"KEH-Rep10-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep1-7KO-HEK293T-Cyto-BS_star_realigned\",\n",
    "               \"KEH-Rep5-7KO-HEK293T-Cyto-BS_star_realigned\"]\n",
    "rep_matches = re.findall(r\"Rep\\d+\", str(folder_list))\n",
    "rep_list = sorted(set(rep_matches), key = lambda x: int(x[3:]))\n",
    "\n",
    "print(\"\\nNOW TESTING REP_LIST\")\n",
    "print(f\"Expected: ['Rep1', 'Rep5', 'Rep10'], Actual: {rep_list}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "41a0d4a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING KEY VARIABLE\n",
      "---\n",
      "1. Recall that subfolder: KEH-Rep2-7KO-HEK293T-Cyto-BS_star_realigned\n",
      "2. From here, we select out Rep1 and BS\n",
      "3. Expected format: Rep1_(A|C|G|T|Deletions|DeletionRate|RealRate|TotalCoverage)_BS\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'A': 'Rep2_A_BS',\n",
       " 'C': 'Rep2_C_BS',\n",
       " 'G': 'Rep2_G_BS',\n",
       " 'T': 'Rep2_T_BS',\n",
       " 'Deletions': 'Rep2_Deletions_BS',\n",
       " 'DeletionRate': 'Rep2_DeletionRate_BS',\n",
       " 'RealRate': 'Rep2_RealRate_BS',\n",
       " 'TotalCoverage': 'Rep2_TotalCoverage_BS'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key = {base_key: make_key(subfolder, base_key) for base_key \n",
    "       in [\"A\", \"C\", \"G\", \"T\", \"Deletions\", \"DeletionRate\", \"RealRate\", \"TotalCoverage\"]}\n",
    "\n",
    "print(\"NOW TESTING KEY VARIABLE\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "print(f\"1. Recall that subfolder: {subfolder}\",\n",
    "      \"2. From here, we select out Rep1 and BS\",\n",
    "      f\"3. Expected format: Rep1_(A|C|G|T|Deletions|DeletionRate|RealRate|TotalCoverage)_BS\", sep = \"\\n\")\n",
    "key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e69ee094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "      <th>Rep2_DeletionRate_BS</th>\n",
       "      <th>Rep2_RealRate_BS</th>\n",
       "      <th>Rep2_TotalCoverage_BS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>249</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  Rep2_Deletions_BS  \\\n",
       "0        249          0          0          0                 24   \n",
       "\n",
       "   Rep2_DeletionRate_BS  Rep2_RealRate_BS  Rep2_TotalCoverage_BS  \n",
       "0              0.087912          0.506265                    273  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Dynamically rename columns in df_clean\n",
    "df_final = df_clean.rename(columns = {\"A\": key[\"A\"], \n",
    "                                      \"C\": key[\"C\"], \n",
    "                                      \"G\": key[\"G\"], \n",
    "                                      \"T\": key[\"T\"], \n",
    "                                      \"Deletions\": key[\"Deletions\"], \n",
    "                                      \"DeletionRate\": key[\"DeletionRate\"], \n",
    "                                      \"RealRate\": key[\"RealRate\"],\n",
    "                                      \"TotalCoverage\": key[\"TotalCoverage\"]})\n",
    "\n",
    "## Verify that it worked\n",
    "diff_cols = df_final.columns.difference(constant_cols, sort = False)\n",
    "df_final[diff_cols[3:]].head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c52c92",
   "metadata": {},
   "source": [
    "### Testing get_sample_group()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b0cb1263",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function definition\n",
    "def get_sample_group(folder_name):\n",
    "    \"\"\"\n",
    "    PURPOSE:\n",
    "    Given input folder names, extract the group name\n",
    "    by returning the first capture group in RegEx.\n",
    "    ---\n",
    "    EXAMPLE: \n",
    "    '7KO-Cyto-BS_processed_fastqs' -> '7KO-Cyto'\n",
    "    \"\"\"\n",
    "    match = re.match(r\"(.+)-(?:BS|NBS)_processed_fastqs\", folder_name)\n",
    "    return match.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "38256626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING GET_SAMPLE_GROUP()\n",
      "Recall that folder_name: 7KO-Cyto-BS_processed_fastqs\n",
      "---\n",
      "Expected: 7KO-Cyto, Actual: 7KO-Cyto\n",
      "Expected: 7KO-Cyto, Actual: 7KO-Cyto\n",
      "Expected: WT-Nuc, Actual: WT-Nuc\n",
      "Expected: WT-Nuc, Actual: WT-Nuc\n"
     ]
    }
   ],
   "source": [
    "print(\"NOW TESTING GET_SAMPLE_GROUP()\",\n",
    "      f\"Recall that folder_name: {folder_name}\", \n",
    "      \"---\",sep = \"\\n\")\n",
    "print(\"Expected: 7KO-Cyto, Actual:\", get_sample_group(folder_name))\n",
    "print(\"Expected: 7KO-Cyto, Actual:\", get_sample_group(\"7KO-Cyto-NBS_processed_fastqs\"))\n",
    "print(\"Expected: WT-Nuc, Actual:\", get_sample_group(\"WT-Nuc-BS_processed_fastqs\"))\n",
    "print(\"Expected: WT-Nuc, Actual:\", get_sample_group(\"WT-Nuc-NBS_processed_fastqs\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac3f6f69",
   "metadata": {},
   "source": [
    "### Testing filtering & cutoffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ccf9783b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING COLUMN NAME RETRIEVAL\n",
      "---\n",
      "Expected: Rep2_DeletionRate_BS\n",
      "Actual: Rep2_DeletionRate_BS\n",
      "\n",
      "Expected: Rep2_RealRate_BS\n",
      "Actual: Rep2_RealRate_BS\n"
     ]
    }
   ],
   "source": [
    "## Filter by DeletionRate based on keywords in: \n",
    "## (a) folder_name\n",
    "## (b) DeletionRate column (accessed via key)\n",
    "\"\"\"\n",
    "WT:\n",
    "* BS files must have DeletionRate values of >= 0.8\n",
    "* NBS files must have DeletionRate values of <= 0.1\n",
    "---\n",
    "Mutation (PUS7KO):\n",
    "* BS files must have DeletionRate values of <= 0.1\n",
    "\"\"\"\n",
    "print(\"NOW TESTING COLUMN NAME RETRIEVAL\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "dr_pattern = key[\"DeletionRate\"]\n",
    "print(\"Expected: Rep2_DeletionRate_BS\", \n",
    "      f\"Actual: {dr_pattern}\", sep = \"\\n\")\n",
    "\n",
    "rr_pattern = key[\"RealRate\"]\n",
    "print(\"\\nExpected: Rep2_RealRate_BS\", \n",
    "      f\"Actual: {rr_pattern}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a16701da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING REALRATE FILTERING\n",
      "---\n",
      "Size of original df: 35\n",
      "Size of kept_rr: 35\n",
      "\n",
      "Size of original df: 35\n",
      "Size of dropped_rr: 0\n",
      "\n",
      "Expected size of dropped_rr: 0\n",
      "Actual: 0\n"
     ]
    }
   ],
   "source": [
    "## Testing RealRate filtering\n",
    "print(\"NOW TESTING REALRATE FILTERING\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "kept_rr = df_final[df_final[rr_pattern].ge(0.3)]\n",
    "print(f\"Size of original df: {len(df_clean)}\",\n",
    "      f\"Size of kept_rr: {len(kept_rr)}\", sep = \"\\n\")\n",
    "\n",
    "dropped_rr = df_final[df_final[rr_pattern].lt(0.3)]\n",
    "print(f\"\\nSize of original df: {len(df_clean)}\",\n",
    "      f\"Size of dropped_rr: {len(dropped_rr)}\", sep = \"\\n\")\n",
    "\n",
    "print(f\"\\nExpected size of dropped_rr: {len(df_clean) - len(kept_rr)}\",\n",
    "      f\"Actual: {len(dropped_rr)}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b1766b2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING COVERAGE FILTERING\n",
      "---\n",
      "Size of kept_rr: 35\n",
      "Size of kept_cov: 35\n"
     ]
    }
   ],
   "source": [
    "## Testing coverage filtering\n",
    "print(\"NOW TESTING COVERAGE FILTERING\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "coverage_list = [col for col in kept_rr.columns \n",
    "                 if re.search(\"(A|C|G|T|Deletions)_.*\", col)]\n",
    "kept_rr[\"TotalCoverage\"] = kept_rr[coverage_list].sum(axis = 1)\n",
    "kept_cov = kept_rr[kept_rr[\"TotalCoverage\"].ge(20)]\n",
    "\n",
    "print(f\"Size of kept_rr: {len(kept_rr)}\",\n",
    "      f\"Size of kept_cov: {len(kept_cov)}\", sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8fe10a35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING DELETIONRATE FILTERING\n",
      "---\n",
      "Size of original df: 35\n",
      "Size of kept_rr: 35\n",
      "Size of kept_cov: 35\n",
      "Size of final filtered df: 35\n",
      "\n",
      "Total rows removed: 0\n",
      "The dataset shrunk by approximately 0% after filtering\n"
     ]
    }
   ],
   "source": [
    "## Testing DeletionRate filtering\n",
    "print(\"NOW TESTING DELETIONRATE FILTERING\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "if re.match(fr\"(WT|7KO).*\", str(folder_name)):\n",
    "    if re.match(fr\"WT.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            df_final = kept_cov[kept_cov[dr_pattern].ge(0.8)]\n",
    "        else: \n",
    "            df_final = kept_cov[kept_cov[dr_pattern].le(0.1)]\n",
    "\n",
    "    if re.match(fr\"7KO.*\", str(folder_name)):\n",
    "        if \"_BS\" in dr_pattern:\n",
    "            ## THIS LINE SHOULD EXECUTE\n",
    "            ## Recall: folder_name = 7KO-Cyto-BS_processed_fastqs\n",
    "            df_final = kept_cov[kept_cov[dr_pattern].le(0.1)]\n",
    "\n",
    "    df_final = df_final.sort_values(by = dr_pattern, ascending = False)\n",
    "\n",
    "print(f\"Size of original df: {len(df_clean)}\",\n",
    "      f\"Size of kept_rr: {len(kept_rr)}\",\n",
    "      f\"Size of kept_cov: {len(kept_cov)}\",\n",
    "      f\"Size of final filtered df: {len(df_final)}\", sep = \"\\n\")\n",
    "\n",
    "rows_removed = (len(df_clean) - len(df_final))\n",
    "percent_change = round((rows_removed/len(df_clean)) * 100)\n",
    "print(f\"\\nTotal rows removed: {rows_removed}\",\n",
    "      f\"The dataset shrunk by approximately {percent_change}% after filtering\",\n",
    "      sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfa21cc4",
   "metadata": {},
   "source": [
    "### Investigating results of df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2abbe280",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TranscriptPosStart</th>\n",
       "      <th>TranscriptPosEnd</th>\n",
       "      <th>TranscriptModBase</th>\n",
       "      <th>GenomicModBase</th>\n",
       "      <th>TranscriptLength</th>\n",
       "      <th>...</th>\n",
       "      <th>fit_b</th>\n",
       "      <th>Rep2_A_BS</th>\n",
       "      <th>Rep2_C_BS</th>\n",
       "      <th>Rep2_G_BS</th>\n",
       "      <th>Rep2_T_BS</th>\n",
       "      <th>Rep2_Deletions_BS</th>\n",
       "      <th>Rep2_DeletionRate_BS</th>\n",
       "      <th>Rep2_RealRate_BS</th>\n",
       "      <th>Rep2_TotalCoverage_BS</th>\n",
       "      <th>TotalCoverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-XM_005245453.2</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3857</td>\n",
       "      <td>3862</td>\n",
       "      <td>3859</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_022731.5</td>\n",
       "      <td>UUUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>-</td>\n",
       "      <td>3860</td>\n",
       "      <td>3865</td>\n",
       "      <td>3862</td>\n",
       "      <td>205715358</td>\n",
       "      <td>6399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>249</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.506265</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_012458.4</td>\n",
       "      <td>UAUAG</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>-</td>\n",
       "      <td>1250</td>\n",
       "      <td>1255</td>\n",
       "      <td>1252</td>\n",
       "      <td>2426036</td>\n",
       "      <td>1664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.735916</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001322267.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1730</td>\n",
       "      <td>1735</td>\n",
       "      <td>1732</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_004697.5</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1691</td>\n",
       "      <td>1696</td>\n",
       "      <td>1693</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rna-NM_001322266.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1727</td>\n",
       "      <td>1732</td>\n",
       "      <td>1729</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2936</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rna-NM_001244926.2</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>+</td>\n",
       "      <td>1688</td>\n",
       "      <td>1693</td>\n",
       "      <td>1690</td>\n",
       "      <td>113291699</td>\n",
       "      <td>2897</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.426971</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rna-XM_017021870.3</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>-</td>\n",
       "      <td>4590</td>\n",
       "      <td>4595</td>\n",
       "      <td>4592</td>\n",
       "      <td>52547741</td>\n",
       "      <td>5289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.417143</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rna-NM_001306195.1</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>-</td>\n",
       "      <td>4645</td>\n",
       "      <td>4650</td>\n",
       "      <td>4647</td>\n",
       "      <td>52547741</td>\n",
       "      <td>5344</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.417143</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rna-NM_006628.6</td>\n",
       "      <td>UCUAA</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>-</td>\n",
       "      <td>4667</td>\n",
       "      <td>4672</td>\n",
       "      <td>4669</td>\n",
       "      <td>52547741</td>\n",
       "      <td>5366</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.417143</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TranscriptID  Motif Region         Chrom Strand  TranscriptPosStart  \\\n",
       "0   rna-XM_005245453.2  UUUAG   3UTR  NC_000001.11      -                3857   \n",
       "1      rna-NM_022731.5  UUUAG   3UTR  NC_000001.11      -                3860   \n",
       "2      rna-NM_012458.4  UAUAG   3UTR  NC_000019.10      -                1250   \n",
       "3   rna-NM_001322267.2  UCUAA   3UTR  NC_000009.12      +                1730   \n",
       "4      rna-NM_004697.5  UCUAA   3UTR  NC_000009.12      +                1691   \n",
       "5   rna-NM_001322266.2  UCUAA   3UTR  NC_000009.12      +                1727   \n",
       "6   rna-NM_001244926.2  UCUAA   3UTR  NC_000009.12      +                1688   \n",
       "12  rna-XM_017021870.3  UCUAA   3UTR  NC_000015.10      -                4590   \n",
       "15  rna-NM_001306195.1  UCUAA   3UTR  NC_000015.10      -                4645   \n",
       "14     rna-NM_006628.6  UCUAA   3UTR  NC_000015.10      -                4667   \n",
       "\n",
       "    TranscriptPosEnd  TranscriptModBase  GenomicModBase  TranscriptLength  \\\n",
       "0               3862               3859       205715358              6396   \n",
       "1               3865               3862       205715358              6399   \n",
       "2               1255               1252         2426036              1664   \n",
       "3               1735               1732       113291699              2939   \n",
       "4               1696               1693       113291699              2900   \n",
       "5               1732               1729       113291699              2936   \n",
       "6               1693               1690       113291699              2897   \n",
       "12              4595               4592        52547741              5289   \n",
       "15              4650               4647        52547741              5344   \n",
       "14              4672               4669        52547741              5366   \n",
       "\n",
       "    ...  fit_b  Rep2_A_BS  Rep2_C_BS  Rep2_G_BS  Rep2_T_BS  Rep2_Deletions_BS  \\\n",
       "0   ...  0.001        249          0          0          0                 24   \n",
       "1   ...  0.001        249          0          0          0                 24   \n",
       "2   ...  0.001          8         15          0          0                  2   \n",
       "3   ...  0.001          0          0          0         25                  2   \n",
       "4   ...  0.001          0          0          0         25                  2   \n",
       "5   ...  0.001          0          0          0         25                  2   \n",
       "6   ...  0.001          0          0          0         25                  2   \n",
       "12  ...  0.001         26          0          0          0                  2   \n",
       "15  ...  0.001         26          0          0          0                  2   \n",
       "14  ...  0.001         26          0          0          0                  2   \n",
       "\n",
       "    Rep2_DeletionRate_BS  Rep2_RealRate_BS  Rep2_TotalCoverage_BS  \\\n",
       "0               0.087912          0.506265                    273   \n",
       "1               0.087912          0.506265                    273   \n",
       "2               0.080000          0.735916                     25   \n",
       "3               0.074074          0.426971                     27   \n",
       "4               0.074074          0.426971                     27   \n",
       "5               0.074074          0.426971                     27   \n",
       "6               0.074074          0.426971                     27   \n",
       "12              0.071429          0.417143                     28   \n",
       "15              0.071429          0.417143                     28   \n",
       "14              0.071429          0.417143                     28   \n",
       "\n",
       "    TotalCoverage  \n",
       "0             273  \n",
       "1             273  \n",
       "2              25  \n",
       "3              27  \n",
       "4              27  \n",
       "5              27  \n",
       "6              27  \n",
       "12             28  \n",
       "15             28  \n",
       "14             28  \n",
       "\n",
       "[10 rows x 26 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9ad79f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "original_motifs = df_clean[\"Motif\"].unique()\n",
    "final_motifs = df_final[\"Motif\"].unique()\n",
    "set_diff = list(set(original_motifs) - set(final_motifs))\n",
    "\n",
    "if set_diff:\n",
    "    print(f\"The only UNUAR motif missing in the final dataframe is {set_diff[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c80bdcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## For priority_filtered\n",
    "# df_final = df.drop(columns = constant_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e81c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find counts for each UNUAR motif\n",
    "# df[\"Motif\"].groupby()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16c837f",
   "metadata": {},
   "source": [
    "### Verifying DeletionRate and RealRate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d96699ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING DELETIONRATE CALCULATIONS\n",
      "---\n",
      "Expected total sum: 273\n",
      "Actual total sum: 273\n",
      "Expected deletion ct: 24\n",
      "Actual deletion ct: 24\n",
      "Expected deletion rate: 0.08791208791208792\n",
      "Actual deletion rate: 0.0879120879120879\n",
      "\n",
      "Deletion rate is correct!\n"
     ]
    }
   ],
   "source": [
    "## Testing DeletionRate calculations\n",
    "print(\"NOW TESTING DELETIONRATE CALCULATIONS\",\n",
    "      \"---\", sep = \"\\n\")\n",
    "first_row = df_final[diff_cols].head(1)\n",
    "first_row\n",
    "\n",
    "count_cols = diff_cols[3:8].to_list()\n",
    "total_sum = int(first_row[count_cols].sum().sum())\n",
    "manual_sum = 249 + 0 + 0 + 0 + 24\n",
    "print(f\"Expected total sum: {manual_sum}\", \n",
    "      f\"Actual total sum: {total_sum}\", sep = \"\\n\")\n",
    "\n",
    "deletion_col = [col for col in diff_cols if re.search(\"Deletions.*\", col)]\n",
    "deletion_ct = int(first_row[deletion_col[0]].iloc[0])\n",
    "deletion_rate = total_sum\n",
    "print(f\"Expected deletion ct: 24\",\n",
    "      f\"Actual deletion ct: {deletion_ct}\", sep = \"\\n\")\n",
    "\n",
    "dr_col = [col for col in diff_cols if re.search(\"DeletionRate.*\", col)]\n",
    "expected_dr = float(deletion_ct/total_sum)\n",
    "actual_dr = float(first_row[dr_col[0]].iloc[0])\n",
    "print(f\"Expected deletion rate: {expected_dr}\",\n",
    "      f\"Actual deletion rate: {actual_dr}\", sep = \"\\n\")\n",
    "\n",
    "if round(expected_dr, 5) == round(actual_dr, 5):\n",
    "    print(\"\\nDeletion rate is correct!\")\n",
    "else:\n",
    "    print(\"\\nDeletion rate not calculated correctly\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4c85b29d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOW TESTING REALRATE CALCULATIONS\n",
      "---\n",
      "Expected real rate: 0.5062647045046491\n",
      "Actual real rate: 0.5062647045046491\n",
      "\n",
      "Real rate is correct!\n"
     ]
    }
   ],
   "source": [
    "print(\"NOW TESTING REALRATE CALCULATIONS\", \n",
    "      \"---\", sep = \"\\n\")\n",
    "rr_col = [col for col in diff_cols if re.search(\"RealRate\", col)]\n",
    "actual_rr = first_row[rr_col[0]].iloc[0]\n",
    "\n",
    "fit_c = first_row[\"fit_c\"].iloc[0]\n",
    "fit_s = first_row[\"fit_s\"].iloc[0]\n",
    "fit_b = first_row[\"fit_b\"].iloc[0]\n",
    "num = fit_b - expected_dr\n",
    "denom = fit_c * (fit_b + fit_s\n",
    "                 - fit_s * expected_dr \n",
    "                 - 1)\n",
    "expected_rr = num/denom\n",
    "\n",
    "print(f\"Expected real rate: {expected_rr}\",\n",
    "      f\"Actual real rate: {actual_rr}\", sep = \"\\n\")\n",
    "\n",
    "if round(expected_rr, 5) == round(actual_rr, 5):\n",
    "    print(\"\\nReal rate is correct!\")\n",
    "else: \n",
    "    print(\"\\nReal rate not calculated correctly\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
